synapse/handlers/federation_event.py

```diff
@@ -542,7 +542,11 @@ async def backfill(
                     f"room {ev.room_id}, when we were backfilling in {room_id}"
```

Comment:
Good shout on the direction, fixed that up in https://github.com/matrix-org/synapse/pull/11114/commits/fb8e2814f1ce78ec0c3df719d2b64ddaa1eb7f4d.

Still contemplating whether this is too confusing.

Reply:
Here are the problem scenarios:


### Scenario 1: Partial at the bottom of the historical branch

#### Expected sort

```
backfill sorted_events=[
  "event_id=$c314gfzp5eK3Ti7CFADPe5DV31ew4P-Sam_tLd70QB4,depth=11,body=Message 1 (eventIDsAfter),prevs=['$Tw6QXdermFoW4JQcc7Zu5U3Khlz1qVG6BD6Vq6VmiUY']",
  "event_id=$Tw6QXdermFoW4JQcc7Zu5U3Khlz1qVG6BD6Vq6VmiUY,depth=10,body=Message 0 (eventIDsAfter),prevs=['$PVJJWg-rYE1a3TLEAQG-1HexD-b6QAFnlgJjdJooF8Y']",
  "event_id=$_ktfy1Mzf3HF2KtabjoK48pDiBcJ2CjUWya6DagKxvg,depth=10,body=org.matrix.msc2716.insertion,prevs=['$PVJJWg-rYE1a3TLEAQG-1HexD-b6QAFnlgJjdJooF8Y']",
  "event_id=$Ki9MNoPjmZe0VwseDb82USns9HpE2T99FLoB8dPaQ4U,depth=10,body=org.matrix.msc2716.batch,prevs=['$3v5SvCqrBH-jYaE9IZGuF8auc9qFU6HRx_WdZMVVFWM']",
  "event_id=$3v5SvCqrBH-jYaE9IZGuF8auc9qFU6HRx_WdZMVVFWM,depth=10,body=Historical 99 (batch=0),prevs=['$BxVwvCHVWLzrWU1XfH7E6ZQo-x783GkySh3xx462YT4']",
  "event_id=$BxVwvCHVWLzrWU1XfH7E6ZQo-x783GkySh3xx462YT4,depth=10,body=Historical 98 (batch=0),prevs=['$yVdBlIMWDDesTVEYnfcwQ0Y6SDGB9ZwoU5P56pmhbi4']",
  "event_id=$yVdBlIMWDDesTVEYnfcwQ0Y6SDGB9ZwoU5P56pmhbi4,depth=10,body=Historical 97 (batch=0),prevs=['$Hol5KY-QBYzFcqkLNbxkYg9ZDYgj9Vo-pqOhsNb_bi4']",
  "event_id=$Hol5KY-QBYzFcqkLNbxkYg9ZDYgj9Vo-pqOhsNb_bi4,depth=10,body=Historical 96 (batch=0),prevs=['$xm6Ylbk4rTXrZaDZ31yrOwE-XparM4cyq41VlavlgIc']",
  "event_id=$xm6Ylbk4rTXrZaDZ31yrOwE-XparM4cyq41VlavlgIc,depth=10,body=Historical 95 (batch=0),prevs=['$2poYyo6Dao9Cvt5oJCvQO2RCWeYkKKwXOuPtcCLOqZI']",
  "event_id=$2poYyo6Dao9Cvt5oJCvQO2RCWeYkKKwXOuPtcCLOqZI,depth=10,body=Historical 94 (batch=0),prevs=['$019DNNmpib0T8t4_LAkmT3KNh70jyOU2iMJGOkScF_k']",
  "event_id=$019DNNmpib0T8t4_LAkmT3KNh70jyOU2iMJGOkScF_k,depth=10,body=Historical 93 (batch=0),prevs=['$ah4V0y53ApkH1K3X9uqUZnWIHMEzWob-d-PKO1E0LIY']",
  "event_id=$ah4V0y53ApkH1K3X9uqUZnWIHMEzWob-d-PKO1E0LIY,depth=10,body=Historical 92 (batch=0),prevs=['$s8WMqgb4DM93Gn5jgaDumnDF5wG2ufO7cgvuJSZbjkc']",
  "event_id=$s8WMqgb4DM93Gn5jgaDumnDF5wG2ufO7cgvuJSZbjkc,depth=10,body=Historical 91 (batch=0),prevs=['$vEFpS1Nl4sVsZ-JMtQrb-szRv3l9nCVthIPLH3aj6ZA']",
  "event_id=$vEFpS1Nl4sVsZ-JMtQrb-szRv3l9nCVthIPLH3aj6ZA,depth=10,body=Historical 90 (batch=0),prevs=['$QqhWUjaRT46E4uoZ4DVSJJM98rUrQvbaVO2_olk-2ug']",
  "event_id=$QqhWUjaRT46E4uoZ4DVSJJM98rUrQvbaVO2_olk-2ug,depth=10,body=Historical 89 (batch=0),prevs=['$9kBxYCUe1rMjdei643hz9101Qk7rlPQ0YJTp3iBI_tg']",
  "event_id=$9kBxYCUe1rMjdei643hz9101Qk7rlPQ0YJTp3iBI_tg,depth=10,body=Historical 88 (batch=0),prevs=['$Qt6qVnDowNisAGI_vpmxhKMLJ0VVBYrUZvyUSKnySnQ']",
  "event_id=$Qt6qVnDowNisAGI_vpmxhKMLJ0VVBYrUZvyUSKnySnQ,depth=10,body=Historical 87 (batch=0),prevs=['$qjNikTCNI9t33JEToLQKvfckZ_UhyR4DeQeNIJ1zdjQ']",
  "event_id=$qjNikTCNI9t33JEToLQKvfckZ_UhyR4DeQeNIJ1zdjQ,depth=10,body=Historical 86 (batch=0),prevs=['$-lRQLCGrs0dz_90aZrC1c4MTIC_2nsDrjBC4LpRVEqA']",
  "event_id=$-lRQLCGrs0dz_90aZrC1c4MTIC_2nsDrjBC4LpRVEqA,depth=10,body=Historical 85 (batch=0),prevs=['$luq70jw2plnfbgqIKH-f7zX0fNcM0yF0SNUgQfwR1_4']",
  "event_id=$luq70jw2plnfbgqIKH-f7zX0fNcM0yF0SNUgQfwR1_4,depth=10,body=Historical 84 (batch=0),prevs=['$-NxuKkBK024nE3PaVeuXS5TGNiz5aMUv75kvDgTrmK4']",
  "event_id=$-NxuKkBK024nE3PaVeuXS5TGNiz5aMUv75kvDgTrmK4,depth=10,body=Historical 83 (batch=0),prevs=['$pXqlXQw9LSXboKsJMFoeauTUAJQ8DhdFuoyyj3g2dMM']",
  "event_id=$pXqlXQw9LSXboKsJMFoeauTUAJQ8DhdFuoyyj3g2dMM,depth=10,body=Historical 82 (batch=0),prevs=['$BdwcbXFhSPuPi78PmlMun3vb4fKTR4fEErdZZOTHH1k']",
  "event_id=$BdwcbXFhSPuPi78PmlMun3vb4fKTR4fEErdZZOTHH1k,depth=10,body=Historical 81 (batch=0),prevs=['$40TRl3SFqb3ZblHuABoPDKtmDgRTBfs4krpewIfhHQg']",
  "event_id=$40TRl3SFqb3ZblHuABoPDKtmDgRTBfs4krpewIfhHQg,depth=10,body=Historical 80 (batch=0),prevs=['$Sgto0D5ABMxdsqcaXkKQToKkAE7B6kPRKDLAHur8ttA']",
  "event_id=$Sgto0D5ABMxdsqcaXkKQToKkAE7B6kPRKDLAHur8ttA,depth=10,body=Historical 79 (batch=0),prevs=['$ZKje9Q8NQdGd1TGgbAH3HibFjkeyV9SfVoBC1BTPFzQ']",
  "event_id=$ZKje9Q8NQdGd1TGgbAH3HibFjkeyV9SfVoBC1BTPFzQ,depth=10,body=Historical 78 (batch=0),prevs=['$-jh9Uyp7_sNlxG71GO95omuToYoNV0ryPffoYPH0lwQ']",
  "event_id=$-jh9Uyp7_sNlxG71GO95omuToYoNV0ryPffoYPH0lwQ,depth=10,body=Historical 77 (batch=0),prevs=['$ihhagD2WpZzeBmZ3HY9BFFVua5jMa2eIYHqdG3Gc-xI']",
  "event_id=$ihhagD2WpZzeBmZ3HY9BFFVua5jMa2eIYHqdG3Gc-xI,depth=10,body=Historical 76 (batch=0),prevs=['$hLmo3PC0qex7MQ5d8h7Z4Ze2HJyrgSQwT1xxJoUdlfk']",
  "event_id=$hLmo3PC0qex7MQ5d8h7Z4Ze2HJyrgSQwT1xxJoUdlfk,depth=10,body=Historical 75 (batch=0),prevs=['$NjVn157cod4FkJDVz2l-LLoaFFLml4k1pUIq4PuSBc8']",
  "event_id=$NjVn157cod4FkJDVz2l-LLoaFFLml4k1pUIq4PuSBc8,depth=10,body=Historical 74 (batch=0),prevs=['$_Y4Dva5xVYAuIqDrNpllvhpNWeLxpGYZUS66F2kJ_AA']",
  "event_id=$_Y4Dva5xVYAuIqDrNpllvhpNWeLxpGYZUS66F2kJ_AA,depth=10,body=Historical 73 (batch=0),prevs=['$rrtlUjtMaqsZggSNmIHvK15WUzkrGyIbFUbXDBRT1wE']",
  "event_id=$rrtlUjtMaqsZggSNmIHvK15WUzkrGyIbFUbXDBRT1wE,depth=10,body=Historical 72 (batch=0),prevs=['$DU1XngcegPIcLSJ0p-hXesfKohQi2D65WuiNpGefABY']",
  "event_id=$DU1XngcegPIcLSJ0p-hXesfKohQi2D65WuiNpGefABY,depth=10,body=Historical 71 (batch=0),prevs=['$orT3tv_2vP1smgMN2VUF4OLCyBWmbK0Jmt0k74ZiQvQ']",
  "event_id=$orT3tv_2vP1smgMN2VUF4OLCyBWmbK0Jmt0k74ZiQvQ,depth=10,body=Historical 70 (batch=0),prevs=['$TqkZLzllkXzlkJmi-gOQt7x9pd07OZYjebjwaR8Q8_Y']",
  "event_id=$TqkZLzllkXzlkJmi-gOQt7x9pd07OZYjebjwaR8Q8_Y,depth=10,body=Historical 69 (batch=0),prevs=['$PkBl3XjikUuHJN2HG0V3s_7xdSK8xWHAZR6lnXcmxtk']",
  "event_id=$PkBl3XjikUuHJN2HG0V3s_7xdSK8xWHAZR6lnXcmxtk,depth=10,body=Historical 68 (batch=0),prevs=['$vLSceQHhObd_787W7nGBr7N5fzfyjgtgSDUwsG9EUbM']",
  "event_id=$vLSceQHhObd_787W7nGBr7N5fzfyjgtgSDUwsG9EUbM,depth=10,body=Historical 67 (batch=0),prevs=['$WMOAwKpzmvUsfPA6XQpVJQSynwGEP9tGPyorQitKE9o']",
  "event_id=$WMOAwKpzmvUsfPA6XQpVJQSynwGEP9tGPyorQitKE9o,depth=10,body=Historical 66 (batch=0),prevs=['$gzai9umMkaVsgcXUvAnThyyGWMj76mx7A4h-4uvqDNM']",
  "event_id=$gzai9umMkaVsgcXUvAnThyyGWMj76mx7A4h-4uvqDNM,depth=10,body=Historical 65 (batch=0),prevs=['$n6tMJUwBvRKhskeoe6B1wo-b4bYS-fLWg1I-4iNY9nk']",
  "event_id=$n6tMJUwBvRKhskeoe6B1wo-b4bYS-fLWg1I-4iNY9nk,depth=10,body=Historical 64 (batch=0),prevs=['$UfrsjQy563In-yA3b9Kpo-dLBqYtO8feoVxO0B3_YG4']",
  "event_id=$UfrsjQy563In-yA3b9Kpo-dLBqYtO8feoVxO0B3_YG4,depth=10,body=Historical 63 (batch=0),prevs=['$W5lSrPSpJvY1lCkI7tsEG_gtov1H7MKURjfUmih79eU']",
  "event_id=$W5lSrPSpJvY1lCkI7tsEG_gtov1H7MKURjfUmih79eU,depth=10,body=Historical 62 (batch=0),prevs=['$BhGaW2JeYEbb0TOECWnH_ns5gZE4OYqQJbF7k3w6ygA']",
  "event_id=$BhGaW2JeYEbb0TOECWnH_ns5gZE4OYqQJbF7k3w6ygA,depth=10,body=Historical 61 (batch=0),prevs=['$Q0bGDDQvTB5xYJXhLtR_dFceVqGMCUMNECZyZNG70wI']",
  "event_id=$Q0bGDDQvTB5xYJXhLtR_dFceVqGMCUMNECZyZNG70wI,depth=10,body=Historical 60 (batch=0),prevs=['$SUmDzIs9hG-YB8FIy_HW_AO3MMiqnS_sFzAOA75mODc']",
  "event_id=$SUmDzIs9hG-YB8FIy_HW_AO3MMiqnS_sFzAOA75mODc,depth=10,body=Historical 59 (batch=0),prevs=['$9sLl1w2hU7tl5Rz1nSau68PrxgOhb_cINc3nwAxHBn4']",
  "event_id=$9sLl1w2hU7tl5Rz1nSau68PrxgOhb_cINc3nwAxHBn4,depth=10,body=Historical 58 (batch=0),prevs=['$_x7MurM519Q5S59Lw2B_-atXf_tSaj8qZqqUZRn6P6o']",
  "event_id=$_x7MurM519Q5S59Lw2B_-atXf_tSaj8qZqqUZRn6P6o,depth=10,body=Historical 57 (batch=0),prevs=['$N8wf3rsWbMB069WiWP5_JuBHY62YWbltnV_3fZraIgQ']",
  "event_id=$N8wf3rsWbMB069WiWP5_JuBHY62YWbltnV_3fZraIgQ,depth=10,body=Historical 56 (batch=0),prevs=['$7hTCl76JcG-MRtTGNqXnENZIXHfNnT_v3KXyFS08N4c']",
  "event_id=$7hTCl76JcG-MRtTGNqXnENZIXHfNnT_v3KXyFS08N4c,depth=10,body=Historical 55 (batch=0),prevs=['$EK8g07iJNozuoGvygr6UVwqyscce9zu5iyBPqNcLBxs']",
  "event_id=$EK8g07iJNozuoGvygr6UVwqyscce9zu5iyBPqNcLBxs,depth=10,body=Historical 54 (batch=0),prevs=['$JikjHhjhN1Actljtp18KmFQ8T1DCN9Vzpmcs3g1IPvA']",
  "event_id=$JikjHhjhN1Actljtp18KmFQ8T1DCN9Vzpmcs3g1IPvA,depth=10,body=Historical 53 (batch=0),prevs=['$3_vnUSH2MGaIvopTBp0wEfXkpcJOSZh_w-mIXv6eKJk']",
  "event_id=$3_vnUSH2MGaIvopTBp0wEfXkpcJOSZh_w-mIXv6eKJk,depth=10,body=Historical 52 (batch=0),prevs=['$7oGQwQiBEuL1XbpVzLh4ofeIWC8OoDryefscocbqonE']",
  "event_id=$7oGQwQiBEuL1XbpVzLh4ofeIWC8OoDryefscocbqonE,depth=10,body=Historical 51 (batch=0),prevs=['$KHHHJA7Aana-fqaczESSP1CidOzlPYbe2qIyg8ewF0w']",
  "event_id=$KHHHJA7Aana-fqaczESSP1CidOzlPYbe2qIyg8ewF0w,depth=10,body=Historical 50 (batch=0),prevs=['$adP29CH4kXafp4Pm3Lq4FS7O3V4Qmu9Nka_26y6IpJE']",
  "event_id=$adP29CH4kXafp4Pm3Lq4FS7O3V4Qmu9Nka_26y6IpJE,depth=10,body=Historical 49 (batch=0),prevs=['$ZP0d_kAykGPXKYkg97MZ-tEkYf2X-LJd-Y0S71LmsD8']",
  "event_id=$ZP0d_kAykGPXKYkg97MZ-tEkYf2X-LJd-Y0S71LmsD8,depth=10,body=Historical 48 (batch=0),prevs=['$Ap4LBg6EKecDbOqsIUNarVH8M-sxMfUE4fvnHqi6oEo']",
  "event_id=$Ap4LBg6EKecDbOqsIUNarVH8M-sxMfUE4fvnHqi6oEo,depth=10,body=Historical 47 (batch=0),prevs=['$x2pHAmfPro2CaMGkbNT-ZEHbD1PQaEj6PfHHC5iDRIQ']",
  "event_id=$x2pHAmfPro2CaMGkbNT-ZEHbD1PQaEj6PfHHC5iDRIQ,depth=10,body=Historical 46 (batch=0),prevs=['$FgWKeJRbGkxaSTZGRTJZ0lV4B4irja-_gfJyUDz8M3c']",
  "event_id=$FgWKeJRbGkxaSTZGRTJZ0lV4B4irja-_gfJyUDz8M3c,depth=10,body=Historical 45 (batch=0),prevs=['$fMkY9bmk46l5PyptFHRycBdotFYxLYyA4hgEZF9pjAY']",
  "event_id=$fMkY9bmk46l5PyptFHRycBdotFYxLYyA4hgEZF9pjAY,depth=10,body=Historical 44 (batch=0),prevs=['$g4pVnEe1p2EpbOSougYf8OZgw2CeI9iByEr3uDrSPuY']",
  "event_id=$g4pVnEe1p2EpbOSougYf8OZgw2CeI9iByEr3uDrSPuY,depth=10,body=Historical 43 (batch=0),prevs=['$ojVbpac4lkvgmJo9Mx0ne6FUlo7ACg2LfoGftpUH9ZI']",
  "event_id=$ojVbpac4lkvgmJo9Mx0ne6FUlo7ACg2LfoGftpUH9ZI,depth=10,body=Historical 42 (batch=0),prevs=['$F9Hc7a0A9qxAi_DezzzKUsk13yChWE0d42fSaYBjgPc']",
  "event_id=$F9Hc7a0A9qxAi_DezzzKUsk13yChWE0d42fSaYBjgPc,depth=10,body=Historical 41 (batch=0),prevs=['$pUIxTCg76o77uMfxJjIK2weqM9T6IQ6YekivZDgnD20']",
  "event_id=$pUIxTCg76o77uMfxJjIK2weqM9T6IQ6YekivZDgnD20,depth=10,body=Historical 40 (batch=0),prevs=['$9M9JqLHmJb2VCBBKjkc9Ej5IAwG8v2n4YChD1SCKjFQ']",
  "event_id=$9M9JqLHmJb2VCBBKjkc9Ej5IAwG8v2n4YChD1SCKjFQ,depth=10,body=Historical 39 (batch=0),prevs=['$ib1BeelWRyBqFi9VdEhC153NTdqQpgGLAr07ZJj9-MQ']",
  "event_id=$ib1BeelWRyBqFi9VdEhC153NTdqQpgGLAr07ZJj9-MQ,depth=10,body=Historical 38 (batch=0),prevs=['$61R1kvPM8_Nt8rLTDdEoUAs3Ue7JpiBQmZW9de14iss']",
  "event_id=$61R1kvPM8_Nt8rLTDdEoUAs3Ue7JpiBQmZW9de14iss,depth=10,body=Historical 37 (batch=0),prevs=['$5dXix6QG4o3cZRrXitSAWE2TJ44c_9gz_pF_gQKmIZw']",
  "event_id=$5dXix6QG4o3cZRrXitSAWE2TJ44c_9gz_pF_gQKmIZw,depth=10,body=Historical 36 (batch=0),prevs=['$lEkNauoLWeP2Kc_Z0e_6FjXwDgkbzqSDxhzqYYYRsdg']",
  "event_id=$lEkNauoLWeP2Kc_Z0e_6FjXwDgkbzqSDxhzqYYYRsdg,depth=10,body=Historical 35 (batch=0),prevs=['$2BfdvXN-u6rj4JFgPP1bcr3JE9-ur7JcF7ihAYzYHUg']",
  "event_id=$2BfdvXN-u6rj4JFgPP1bcr3JE9-ur7JcF7ihAYzYHUg,depth=10,body=Historical 34 (batch=0),prevs=['$nG-fNtLkJHFSR38nZtijeabE3uYEAOuvq2HP0bUGyJk']",
  "event_id=$nG-fNtLkJHFSR38nZtijeabE3uYEAOuvq2HP0bUGyJk,depth=10,body=Historical 33 (batch=0),prevs=['$fAZc_twttB_9XybV9ajjtajqe-76chAdSdHjcz6Tfb8']",
  "event_id=$fAZc_twttB_9XybV9ajjtajqe-76chAdSdHjcz6Tfb8,depth=10,body=Historical 32 (batch=0),prevs=['$Rem6erJTsvQcpgfzWF5ia_LCwmXM7OLWvDxMltu3RB4']",
  "event_id=$Rem6erJTsvQcpgfzWF5ia_LCwmXM7OLWvDxMltu3RB4,depth=10,body=Historical 31 (batch=0),prevs=['$SQN4Q0owMx2gU-4J-qD88Tz3VXpCpLaz9n0O1bIa1kU']",
  "event_id=$SQN4Q0owMx2gU-4J-qD88Tz3VXpCpLaz9n0O1bIa1kU,depth=10,body=Historical 30 (batch=0),prevs=['$sEasprLY-iLBCFLWbZxv-MLSEaOBN0NGihSD6MM68X4']",
  "event_id=$sEasprLY-iLBCFLWbZxv-MLSEaOBN0NGihSD6MM68X4,depth=10,body=Historical 29 (batch=0),prevs=['$V8rf_7aw2xehL8qOZIwcpJfWOmphrvMTDMoFtrs950Y']",
  "event_id=$V8rf_7aw2xehL8qOZIwcpJfWOmphrvMTDMoFtrs950Y,depth=10,body=Historical 28 (batch=0),prevs=['$6BB4zKQJA6Lwz4b5xKPRuLs_ywV1dvQrAvmBCqL2YDw']",
  "event_id=$6BB4zKQJA6Lwz4b5xKPRuLs_ywV1dvQrAvmBCqL2YDw,depth=10,body=Historical 27 (batch=0),prevs=['$4AGk1DlNQhaTesircrVUXBZFDWNBq1Lut_dLJbn82JU']",
  "event_id=$4AGk1DlNQhaTesircrVUXBZFDWNBq1Lut_dLJbn82JU,depth=10,body=Historical 26 (batch=0),prevs=['$_cNNXIC9olgpnudDkug-ktUSKiZM71HvnDgPz0sWXfE']",
  "event_id=$_cNNXIC9olgpnudDkug-ktUSKiZM71HvnDgPz0sWXfE,depth=10,body=Historical 25 (batch=0),prevs=['$0ifr6p9nnJ3BCcOlxItaE8AcLdybZ5Urx-ivUJjaXHA']",
  "event_id=$0ifr6p9nnJ3BCcOlxItaE8AcLdybZ5Urx-ivUJjaXHA,depth=10,body=Historical 24 (batch=0),prevs=['$-8-I_FslO7Gvj-PL3pM9ogEgRjYV2aBrY9rLbDCgC2k']",
  "event_id=$-8-I_FslO7Gvj-PL3pM9ogEgRjYV2aBrY9rLbDCgC2k,depth=10,body=Historical 23 (batch=0),prevs=['$9r34krvdKDH9LKhcDjrFWp7daXJDwMupq-um9orXApk']",
  "event_id=$9r34krvdKDH9LKhcDjrFWp7daXJDwMupq-um9orXApk,depth=10,body=Historical 22 (batch=0),prevs=['$z7Fa5Kn9uoQLpC1gjaTbtoye5TdLWv-kx2zdXOLrC54']",
  "event_id=$z7Fa5Kn9uoQLpC1gjaTbtoye5TdLWv-kx2zdXOLrC54,depth=10,body=Historical 21 (batch=0),prevs=['$2uuOUt4M_qqs9mAMJWWK7LXPPB2FYuUal0v5iC06zIY']",
  "event_id=$2uuOUt4M_qqs9mAMJWWK7LXPPB2FYuUal0v5iC06zIY,depth=10,body=Historical 20 (batch=0),prevs=['$PRTtMhVGlUpUAHoT9p2GVnkY2GHNi-To809HFN6TkUc']",
  "event_id=$PRTtMhVGlUpUAHoT9p2GVnkY2GHNi-To809HFN6TkUc,depth=10,body=Historical 19 (batch=0),prevs=['$ANgsRYx4tP4i41cB3ZVOZTsfP3HzzjQyKACmBpAafmI']",
  "event_id=$ANgsRYx4tP4i41cB3ZVOZTsfP3HzzjQyKACmBpAafmI,depth=10,body=Historical 18 (batch=0),prevs=['$K07Lok7DO2mnuDz7rRXWf2XshAMCSkKGq9A0u-5Hp_o']",
  "event_id=$K07Lok7DO2mnuDz7rRXWf2XshAMCSkKGq9A0u-5Hp_o,depth=10,body=Historical 17 (batch=0),prevs=['$wQ6iaNkRY4vMPjzpsV8WNaqoKxyjIrIn3zUZGNtPGls']",
  "event_id=$wQ6iaNkRY4vMPjzpsV8WNaqoKxyjIrIn3zUZGNtPGls,depth=10,body=Historical 16 (batch=0),prevs=['$Z-6TMGWoXxL6USvSNUmv6asnUH4ra-YOUi0O_uNADCM']",
  "event_id=$Z-6TMGWoXxL6USvSNUmv6asnUH4ra-YOUi0O_uNADCM,depth=10,body=Historical 15 (batch=0),prevs=['$U8iKKA7TdOTva0uXFkDOuQZS9xi6GuMewMu6yVL7t-c']",
  "event_id=$U8iKKA7TdOTva0uXFkDOuQZS9xi6GuMewMu6yVL7t-c,depth=10,body=Historical 14 (batch=0),prevs=['$lOory_yjG64n9-CNSpPXcOBdJ2Hp6YHT69qb5aLhsFE']",
  "event_id=$lOory_yjG64n9-CNSpPXcOBdJ2Hp6YHT69qb5aLhsFE,depth=10,body=Historical 13 (batch=0),prevs=['$G3fi25UZRO9eEp-JbAFAmeelosmpKw_ezNY1M2hpxEc']",
  "event_id=$G3fi25UZRO9eEp-JbAFAmeelosmpKw_ezNY1M2hpxEc,depth=10,body=Historical 12 (batch=0),prevs=['$WNFtelKcHvICGdB3uhvz5VhZesHu28smlL7gT_hDlrg']",
  "event_id=$WNFtelKcHvICGdB3uhvz5VhZesHu28smlL7gT_hDlrg,depth=10,body=Historical 11 (batch=0),prevs=['$zgiMckYUJhO2DzDngM11MbNRzPewo6N21rZYb6LR9Ko']",
  "event_id=$zgiMckYUJhO2DzDngM11MbNRzPewo6N21rZYb6LR9Ko,depth=10,body=Historical 10 (batch=0),prevs=['$nnNTSnfupektlvP2WAl2Ofmuzm13pjwGvBky-SbCZPQ']",
  "event_id=$nnNTSnfupektlvP2WAl2Ofmuzm13pjwGvBky-SbCZPQ,depth=10,body=Historical 9 (batch=0),prevs=['$WvzAHtMLWzv1KVtImLt9I7qlDXKjjgFS__CyS5uw2Go']",
  "event_id=$WvzAHtMLWzv1KVtImLt9I7qlDXKjjgFS__CyS5uw2Go,depth=10,body=Historical 8 (batch=0),prevs=['$XOe653Bs1shkL8UyzxnLdvFSQSz39dZfrQnwGJaTHTs']",
  "event_id=$XOe653Bs1shkL8UyzxnLdvFSQSz39dZfrQnwGJaTHTs,depth=10,body=Historical 7 (batch=0),prevs=['$WL7ORFfraUf9pDiRMFOzwXHIkicS2Fg14RBM5JGzkcY']",
  "event_id=$WL7ORFfraUf9pDiRMFOzwXHIkicS2Fg14RBM5JGzkcY,depth=10,body=Historical 6 (batch=0),prevs=['$JeKL-CqsynXjLMunbu5OYYTFE3e_l0BpMHIP-99xOwY']",
  "event_id=$JeKL-CqsynXjLMunbu5OYYTFE3e_l0BpMHIP-99xOwY,depth=10,body=Historical 5 (batch=0),prevs=['$cLXTHBY3xCtzHjwShMqL5HWKbW0Zv9K-wBaoZv6h3ro']",
  "event_id=$cLXTHBY3xCtzHjwShMqL5HWKbW0Zv9K-wBaoZv6h3ro,depth=10,body=Historical 4 (batch=0),prevs=['$dc1-00h9Wt7xWq333J1yC_L3xQvhLhHftoioSbc_WTc']"
]
```


#### `sorted_topological` with no fake edges ❌

There is no edge connection between the historical batch and the live DAG so there is no way we could sort this properly.

[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBCIC0tLS0tLS0tLS0tLS0tLS0tPiBBXG4gICAgZW5kXG4gICAgXG4gICAgc3ViZ3JhcGggYmF0Y2gwXG4gICAgICAgIGJhdGNoMC1iYXRjaFtbXCJiYXRjaFwiXV0gLS0-IGJhdGNoMC01KChcIjVcIikpIC0tPiBiYXRjaDAtNCgoXCI0XCIpKSAtLT4gYmF0Y2gwLTMoKFwiM1wiKSkgLS0-IGJhY2tmaWxsRW5kKFwidGhpcyBpcyBhcyBmYXIgYXMgYmFja2ZpbGwgcmVhY2hlZCAoaW1hZ2luZSAxMDAgZXZlbnRzKVwiKVxuICAgIGVuZFxuXG5cblxuICAgIGJhdGNoMC1pbnNlcnRpb25CYXNlWy9pbnNlcnRpb25cXF0gLS0tLS0tLS0tLS0tLS0tPiBBXG5cblxuICAgICUlIGFsaWdubWVudCBsaW5rc1xuICAgIGJhY2tmaWxsRW5kIC0tLS0tLS0tLS0-IEFcbiAgICAlJSBtYWtlIHRoZSBsaW5rcyBpbnZpc2libGUgXG4gICAgbGlua1N0eWxlIDYgc3Ryb2tlLXdpZHRoOjJweCxmaWxsOm5vbmUsc3Ryb2tlOm5vbmU7XG4iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)
![](https://user-images.githubusercontent.com/558581/139520438-5279df68-2034-42ae-a665-85e88f2a0767.png)


```
backfill sorted_events=["event_id=$zxJ_Y6y7pHz1xCGBB2NSarzYSEfjRdww1rxU1HRyl68,depth=10,body=org.matrix.msc2716.batch,prevs=['$CoS0HjBSAMvVSMjJZZMU014qZTFcdGhXfjMRfnv6OSM']",
  "event_id=$CoS0HjBSAMvVSMjJZZMU014qZTFcdGhXfjMRfnv6OSM,depth=10,body=Historical 99 (batch=0),prevs=['$AAdfklEy3X5zjx_woupdjoQg1j8aOAayDZbbiqkGjHc']",
  "event_id=$AAdfklEy3X5zjx_woupdjoQg1j8aOAayDZbbiqkGjHc,depth=10,body=Historical 98 (batch=0),prevs=['$Q44TxTjSEKHBnEhhXtwcfw4_UeMFx88pFwN4_rCtlBI']",
  "event_id=$Q44TxTjSEKHBnEhhXtwcfw4_UeMFx88pFwN4_rCtlBI,depth=10,body=Historical 97 (batch=0),prevs=['$RuAyX8vGyuu1foaqIUo8fmv_nBx-lSUJ_kJsSKhv0PY']",
  "event_id=$RuAyX8vGyuu1foaqIUo8fmv_nBx-lSUJ_kJsSKhv0PY,depth=10,body=Historical 96 (batch=0),prevs=['$RkFqWjsAqhjBdmgiCMnasgUcXtq3w1NBbAYrV4PXwv4']",
  "event_id=$RkFqWjsAqhjBdmgiCMnasgUcXtq3w1NBbAYrV4PXwv4,depth=10,body=Historical 95 (batch=0),prevs=['$03iQTV2meLw1bdEpUXK29nKXBOkUDG6txXlmehSvRVc']",
  "event_id=$03iQTV2meLw1bdEpUXK29nKXBOkUDG6txXlmehSvRVc,depth=10,body=Historical 94 (batch=0),prevs=['$OBuQkw2E3OnbatrHWjbGxbW01xDOEBq7C-cpS_BCj-c']",
  "event_id=$OBuQkw2E3OnbatrHWjbGxbW01xDOEBq7C-cpS_BCj-c,depth=10,body=Historical 93 (batch=0),prevs=['$RJSP1-Ws6JDiG5iLuNBG0am68Lz6jX8w9pBM25BbKcM']",
  "event_id=$RJSP1-Ws6JDiG5iLuNBG0am68Lz6jX8w9pBM25BbKcM,depth=10,body=Historical 92 (batch=0),prevs=['$KliBd_Vw7SuXvFaTIhYlBSaAzRxcDUD_Ga32QuBs2zc']",
  "event_id=$KliBd_Vw7SuXvFaTIhYlBSaAzRxcDUD_Ga32QuBs2zc,depth=10,body=Historical 91 (batch=0),prevs=['$GYgnfIRpa2MVPmaUmHCvBva-GQBg8TTdGHaue10iYvs']",
  "event_id=$GYgnfIRpa2MVPmaUmHCvBva-GQBg8TTdGHaue10iYvs,depth=10,body=Historical 90 (batch=0),prevs=['$xHBR1jdzY5HtVXZc596QfiWEk90m4tUG6_X6z54CNuc']",
  "event_id=$xHBR1jdzY5HtVXZc596QfiWEk90m4tUG6_X6z54CNuc,depth=10,body=Historical 89 (batch=0),prevs=['$BhbUJzRJSzpuAYHsLSoNW97W4I9VSL6hCS899Espb8M']",
  "event_id=$BhbUJzRJSzpuAYHsLSoNW97W4I9VSL6hCS899Espb8M,depth=10,body=Historical 88 (batch=0),prevs=['$fsng6nAUp7JI_EiHKr3yfRz8TclcSJpbmrXrWIuyjk8']",
  "event_id=$fsng6nAUp7JI_EiHKr3yfRz8TclcSJpbmrXrWIuyjk8,depth=10,body=Historical 87 (batch=0),prevs=['$KPuhGCHudEQ2NKFihwzlTwn0qdxjT1WBUh-ZMBCUNJc']",
  "event_id=$KPuhGCHudEQ2NKFihwzlTwn0qdxjT1WBUh-ZMBCUNJc,depth=10,body=Historical 86 (batch=0),prevs=['$6ImmOylkGp1E9xXrRR6OUzSywdysHo1fd3c5QCKECss']",
  "event_id=$6ImmOylkGp1E9xXrRR6OUzSywdysHo1fd3c5QCKECss,depth=10,body=Historical 85 (batch=0),prevs=['$3BReDT2Oz1Tpw2D9YTxblCekB3lfsLoibbh-U7rx_8Q']",
  "event_id=$3BReDT2Oz1Tpw2D9YTxblCekB3lfsLoibbh-U7rx_8Q,depth=10,body=Historical 84 (batch=0),prevs=['$6PjjG4q6xikM5ruD8w0ZsgJJzST674HDX2Evi4AfM1g']",
  "event_id=$6PjjG4q6xikM5ruD8w0ZsgJJzST674HDX2Evi4AfM1g,depth=10,body=Historical 83 (batch=0),prevs=['$X5shzfUAk5QWCgLdDkjLPL6yg8AqROQ3_sgxlYJS-XY']",
  "event_id=$X5shzfUAk5QWCgLdDkjLPL6yg8AqROQ3_sgxlYJS-XY,depth=10,body=Historical 82 (batch=0),prevs=['$ieyFs1BjrDuYWUz8jEj71YZDUJAwlL693nZ9wVzOBcA']",
  "event_id=$ieyFs1BjrDuYWUz8jEj71YZDUJAwlL693nZ9wVzOBcA,depth=10,body=Historical 81 (batch=0),prevs=['$2bEC3xmJI6-GWo75dZ7_EeTEoNq4GSdDlrOvHUn4bPk']",
  "event_id=$2bEC3xmJI6-GWo75dZ7_EeTEoNq4GSdDlrOvHUn4bPk,depth=10,body=Historical 80 (batch=0),prevs=['$-aEKtb9x_0FK5mC_xTu_afd915HaV4EfXJrA4npagHQ']",
  "event_id=$-aEKtb9x_0FK5mC_xTu_afd915HaV4EfXJrA4npagHQ,depth=10,body=Historical 79 (batch=0),prevs=['$kP8d3N2pzezqgRamOOCw9PPapXq5r2XCbN43V2115ao']",
  "event_id=$kP8d3N2pzezqgRamOOCw9PPapXq5r2XCbN43V2115ao,depth=10,body=Historical 78 (batch=0),prevs=['$kH6-ZKIhtZD_bRAweUW8LAUS4V4iiw5E02V_9t4vN-M']",
  "event_id=$kH6-ZKIhtZD_bRAweUW8LAUS4V4iiw5E02V_9t4vN-M,depth=10,body=Historical 77 (batch=0),prevs=['$cQs423uZsRPb2gbxgaePsnR-2-Ban7fLpkWfueq4Bgk']",
  "event_id=$cQs423uZsRPb2gbxgaePsnR-2-Ban7fLpkWfueq4Bgk,depth=10,body=Historical 76 (batch=0),prevs=['$Ba-e_tgeEh1lnNUVYANRUtRoT3rXIkil9W5Hu6xAKOY']",
  "event_id=$Ba-e_tgeEh1lnNUVYANRUtRoT3rXIkil9W5Hu6xAKOY,depth=10,body=Historical 75 (batch=0),prevs=['$SMHdxO2OjcTK6Cv0_bwff8JzHxlyco74Zc5j279dOHM']",
  "event_id=$SMHdxO2OjcTK6Cv0_bwff8JzHxlyco74Zc5j279dOHM,depth=10,body=Historical 74 (batch=0),prevs=['$X751cL9RtVcxMvp0T5j9gE4yeOSAarfFVwl8ByTQAwI']",
  "event_id=$X751cL9RtVcxMvp0T5j9gE4yeOSAarfFVwl8ByTQAwI,depth=10,body=Historical 73 (batch=0),prevs=['$KRgbcA_tVQXHN92hwoHycrWdQqK6BpZJx91tEEmjTaI']",
  "event_id=$KRgbcA_tVQXHN92hwoHycrWdQqK6BpZJx91tEEmjTaI,depth=10,body=Historical 72 (batch=0),prevs=['$4TYNeB6hpe8yVZB52A0Yu6fL7h6RBi-aLp49wsRoUtg']",
  "event_id=$4TYNeB6hpe8yVZB52A0Yu6fL7h6RBi-aLp49wsRoUtg,depth=10,body=Historical 71 (batch=0),prevs=['$BiepOr_w68ByPiLd-38nT4qwXAEjqshYd0DjZ1KnizQ']",
  "event_id=$BiepOr_w68ByPiLd-38nT4qwXAEjqshYd0DjZ1KnizQ,depth=10,body=Historical 70 (batch=0),prevs=['$URWj0cxxMkfcU1sehdr5dIM479EpdjdOl-faLfgMgMA']",
  "event_id=$URWj0cxxMkfcU1sehdr5dIM479EpdjdOl-faLfgMgMA,depth=10,body=Historical 69 (batch=0),prevs=['$uCtisVurQYAnxPmTut9o6mjyAITRsvY45eQ6JDOkvp0']",
  "event_id=$uCtisVurQYAnxPmTut9o6mjyAITRsvY45eQ6JDOkvp0,depth=10,body=Historical 68 (batch=0),prevs=['$qMctPrrFu96AuG97ycHwgWKrSzZ4Z-Cr7UBM1VVD1W4']",
  "event_id=$qMctPrrFu96AuG97ycHwgWKrSzZ4Z-Cr7UBM1VVD1W4,depth=10,body=Historical 67 (batch=0),prevs=['$qWBoKmXFGk9uJIHcz18PUGeDFo3bwsf8-s2o5Fm-ceQ']",
  "event_id=$qWBoKmXFGk9uJIHcz18PUGeDFo3bwsf8-s2o5Fm-ceQ,depth=10,body=Historical 66 (batch=0),prevs=['$m08SYS0_r3Rjdmi7mw4vcBxn12sb6ZoZ4MAAUOzpDcs']",
  "event_id=$m08SYS0_r3Rjdmi7mw4vcBxn12sb6ZoZ4MAAUOzpDcs,depth=10,body=Historical 65 (batch=0),prevs=['$cmeLGJVv1oGqPQ5BurwIHmqvvTGD8XQppKGeg93bLhE']",
  "event_id=$cmeLGJVv1oGqPQ5BurwIHmqvvTGD8XQppKGeg93bLhE,depth=10,body=Historical 64 (batch=0),prevs=['$Z2uJWAWD021BhR-qHgsZUGHcKspxZSR58KDLkT6aIJA']",
  "event_id=$Z2uJWAWD021BhR-qHgsZUGHcKspxZSR58KDLkT6aIJA,depth=10,body=Historical 63 (batch=0),prevs=['$wQpaKkNdS8Xm8N4hdKfYTQtDSwC6wa0G0YUu_619NR4']",
  "event_id=$wQpaKkNdS8Xm8N4hdKfYTQtDSwC6wa0G0YUu_619NR4,depth=10,body=Historical 62 (batch=0),prevs=['$w5jZpyoauIuxAjRSSznnGqKlRRE5UVK5HPwSH1OdTAw']",
  "event_id=$w5jZpyoauIuxAjRSSznnGqKlRRE5UVK5HPwSH1OdTAw,depth=10,body=Historical 61 (batch=0),prevs=['$h4yo7Q6ICs3ZIutnMPCRI0-7OW3nWH7fin8vNWmeyxM']",
  "event_id=$h4yo7Q6ICs3ZIutnMPCRI0-7OW3nWH7fin8vNWmeyxM,depth=10,body=Historical 60 (batch=0),prevs=['$iY55UMYh2OiFlc2VxDXAqWnwQ1nzNkurAhdJmCfUd7w']",
  "event_id=$iY55UMYh2OiFlc2VxDXAqWnwQ1nzNkurAhdJmCfUd7w,depth=10,body=Historical 59 (batch=0),prevs=['$VoBp-fs90mBXTftR7bfU2LsfoPEwp3yvIHz1Hz4YvYM']",
  "event_id=$VoBp-fs90mBXTftR7bfU2LsfoPEwp3yvIHz1Hz4YvYM,depth=10,body=Historical 58 (batch=0),prevs=['$H2WraMzNuWscAIAlVoBkfBRHsLCsslsDragVY_Wav6s']",
  "event_id=$H2WraMzNuWscAIAlVoBkfBRHsLCsslsDragVY_Wav6s,depth=10,body=Historical 57 (batch=0),prevs=['$ljai06Li0YuEX1YF3p3l2hOmFLOFQV-Cweda2TnpLdw']",
  "event_id=$ljai06Li0YuEX1YF3p3l2hOmFLOFQV-Cweda2TnpLdw,depth=10,body=Historical 56 (batch=0),prevs=['$uHvLpVM38QyqQM4ghwitvqB7dkB8bXN_7-MuTt5M3N0']",
  "event_id=$uHvLpVM38QyqQM4ghwitvqB7dkB8bXN_7-MuTt5M3N0,depth=10,body=Historical 55 (batch=0),prevs=['$-oM4q1Qn69iRvZzFWIms6xAvJ3ABpOSno0-OX1inkTI']",
  "event_id=$-oM4q1Qn69iRvZzFWIms6xAvJ3ABpOSno0-OX1inkTI,depth=10,body=Historical 54 (batch=0),prevs=['$wZ_E5tpHqQa4REJAJHBKMBTsqdOVbarTa41XCpNvkkw']",
  "event_id=$wZ_E5tpHqQa4REJAJHBKMBTsqdOVbarTa41XCpNvkkw,depth=10,body=Historical 53 (batch=0),prevs=['$OfoedyVsebcxFvZPrycXXwg7AlC6BXX1rs2tuf-4Tr4']",
  "event_id=$OfoedyVsebcxFvZPrycXXwg7AlC6BXX1rs2tuf-4Tr4,depth=10,body=Historical 52 (batch=0),prevs=['$WnbydqrLmfp9rJMAa20Y-R94zAUKIpSVqVXvtyWW4hQ']",
  "event_id=$WnbydqrLmfp9rJMAa20Y-R94zAUKIpSVqVXvtyWW4hQ,depth=10,body=Historical 51 (batch=0),prevs=['$Dee_gy43NTP8SCfx5eEpVbVcaB4zAMyI8_om178Ajm4']",
  "event_id=$Dee_gy43NTP8SCfx5eEpVbVcaB4zAMyI8_om178Ajm4,depth=10,body=Historical 50 (batch=0),prevs=['$y2ixGkYJPh-XrKd1g3BSpdQv6HCG5vobuIx5E7CHtNk']",
  "event_id=$y2ixGkYJPh-XrKd1g3BSpdQv6HCG5vobuIx5E7CHtNk,depth=10,body=Historical 49 (batch=0),prevs=['$fp4j0KWjSG9ItO6e55tWAGl_UoyOkzMTTOJorN9coU0']",
  "event_id=$fp4j0KWjSG9ItO6e55tWAGl_UoyOkzMTTOJorN9coU0,depth=10,body=Historical 48 (batch=0),prevs=['$l485UTBQvN9svQee5xp95b7TcD1vAZDknrKKiHk-fqg']",
  "event_id=$l485UTBQvN9svQee5xp95b7TcD1vAZDknrKKiHk-fqg,depth=10,body=Historical 47 (batch=0),prevs=['$ERkrc5Kn_Hl45OrvSCyJKCgwcUk2UZn8r8xfrUnDOJY']",
  "event_id=$ERkrc5Kn_Hl45OrvSCyJKCgwcUk2UZn8r8xfrUnDOJY,depth=10,body=Historical 46 (batch=0),prevs=['$dJF9Fyl3JYUP2hhAGAb2aH-xCbzKnzHWCVGB-jfbnfI']",
  "event_id=$dJF9Fyl3JYUP2hhAGAb2aH-xCbzKnzHWCVGB-jfbnfI,depth=10,body=Historical 45 (batch=0),prevs=['$WBUWrHiOo5_wd81WbOBpLZ9a17GLF9mPXvLgQ-S9yrk']",
  "event_id=$WBUWrHiOo5_wd81WbOBpLZ9a17GLF9mPXvLgQ-S9yrk,depth=10,body=Historical 44 (batch=0),prevs=['$Ng4_Z7lKQWCILfxNwymL1ul2NxoPhI-yOYtii_FPKv8']",
  "event_id=$Ng4_Z7lKQWCILfxNwymL1ul2NxoPhI-yOYtii_FPKv8,depth=10,body=Historical 43 (batch=0),prevs=['$mi3C9qIj8e4nFHVdCjAArzh16IRazEtaLSCCT2DmxE4']",
  "event_id=$mi3C9qIj8e4nFHVdCjAArzh16IRazEtaLSCCT2DmxE4,depth=10,body=Historical 42 (batch=0),prevs=['$BLmPeaxnb2v0Xx6Wu-MoMBEE8VfYXR9ZPefJtB8KT_c']",
  "event_id=$BLmPeaxnb2v0Xx6Wu-MoMBEE8VfYXR9ZPefJtB8KT_c,depth=10,body=Historical 41 (batch=0),prevs=['$pzKMDVVoufYvlgaaVGmO63A3WlNoQYzMuhiRQW26BNQ']",
  "event_id=$pzKMDVVoufYvlgaaVGmO63A3WlNoQYzMuhiRQW26BNQ,depth=10,body=Historical 40 (batch=0),prevs=['$j09Zl7JZl2BMTEOHZDxarmuKx0XdN1ZyVXUFs99kLGM']",
  "event_id=$j09Zl7JZl2BMTEOHZDxarmuKx0XdN1ZyVXUFs99kLGM,depth=10,body=Historical 39 (batch=0),prevs=['$ajZl-Yuh-DDqUQnX2R-qOiHVRUziBrrmeOM2pA_5BxI']",
  "event_id=$ajZl-Yuh-DDqUQnX2R-qOiHVRUziBrrmeOM2pA_5BxI,depth=10,body=Historical 38 (batch=0),prevs=['$IEwMVK-oLGgOksUzpEg-xtD6f-U8xVOTEnuBwvCU4lU']",
  "event_id=$IEwMVK-oLGgOksUzpEg-xtD6f-U8xVOTEnuBwvCU4lU,depth=10,body=Historical 37 (batch=0),prevs=['$R7uFJUT0x9ecI0u-6L_0z-_xaVsVCsdVAeojdMmSKgU']",
  "event_id=$R7uFJUT0x9ecI0u-6L_0z-_xaVsVCsdVAeojdMmSKgU,depth=10,body=Historical 36 (batch=0),prevs=['$b7drJA3yjqdmRfFYPBPXAKfc747-E6yTcSMcBNv8v9o']",
  "event_id=$b7drJA3yjqdmRfFYPBPXAKfc747-E6yTcSMcBNv8v9o,depth=10,body=Historical 35 (batch=0),prevs=['$G7g3C4vxSUo_fPxoX23jN0s021Utj2gUhchd-MLFNdk']",
  "event_id=$G7g3C4vxSUo_fPxoX23jN0s021Utj2gUhchd-MLFNdk,depth=10,body=Historical 34 (batch=0),prevs=['$GB1zP8cv9HvHUkN2ckszYR9F_44fpN-vCRVdwXzEb-g']",
  "event_id=$GB1zP8cv9HvHUkN2ckszYR9F_44fpN-vCRVdwXzEb-g,depth=10,body=Historical 33 (batch=0),prevs=['$F_n2bVYy5fY5AbdXJKngQfARhhLC1Iwm02Cn-M3H8AI']",
  "event_id=$F_n2bVYy5fY5AbdXJKngQfARhhLC1Iwm02Cn-M3H8AI,depth=10,body=Historical 32 (batch=0),prevs=['$3XTO9PPXE-KO2DMwagIJ7I_mEFIp4AWD9o6LAQhHBso']",
  "event_id=$3XTO9PPXE-KO2DMwagIJ7I_mEFIp4AWD9o6LAQhHBso,depth=10,body=Historical 31 (batch=0),prevs=['$9zbH64QW9jek2tL8NMTjU9PI7LBMPOW2UmiqGyhYMrM']",
  "event_id=$9zbH64QW9jek2tL8NMTjU9PI7LBMPOW2UmiqGyhYMrM,depth=10,body=Historical 30 (batch=0),prevs=['$_4JyznzCyE820TRdggEJ4FISMZP9UpTAkLQdC2rsqrY']",
  "event_id=$_4JyznzCyE820TRdggEJ4FISMZP9UpTAkLQdC2rsqrY,depth=10,body=Historical 29 (batch=0),prevs=['$g_V3UQ8efrlUhlkQ1AFNgAMe3IsFSQ9saBkgdS8gyGk']",
  "event_id=$g_V3UQ8efrlUhlkQ1AFNgAMe3IsFSQ9saBkgdS8gyGk,depth=10,body=Historical 28 (batch=0),prevs=['$3nTrXxQIciU_qEhK_LCyHRRZMJtpPFkaPeCw90AhJWA']",
  "event_id=$3nTrXxQIciU_qEhK_LCyHRRZMJtpPFkaPeCw90AhJWA,depth=10,body=Historical 27 (batch=0),prevs=['$_-KbmY64XE69hi06MoCiGvP0d5sT6e6xGzjtgqMolnI']",
  "event_id=$_-KbmY64XE69hi06MoCiGvP0d5sT6e6xGzjtgqMolnI,depth=10,body=Historical 26 (batch=0),prevs=['$s5CxFAgpBPwZ7BIU0q5eBQouvUCZ5Yvp8hucYB98EEY']",
  "event_id=$s5CxFAgpBPwZ7BIU0q5eBQouvUCZ5Yvp8hucYB98EEY,depth=10,body=Historical 25 (batch=0),prevs=['$qB5ttPe7NwkQ3_gVYx2ziJmKDUSDn4Nb1qOnIWYMPxQ']",
  "event_id=$qB5ttPe7NwkQ3_gVYx2ziJmKDUSDn4Nb1qOnIWYMPxQ,depth=10,body=Historical 24 (batch=0),prevs=['$2PeUavaCDlCkfD856jXj0hRoFF50XfZBHH-4EoHl6qk']",
  "event_id=$2PeUavaCDlCkfD856jXj0hRoFF50XfZBHH-4EoHl6qk,depth=10,body=Historical 23 (batch=0),prevs=['$4cZ4mezAPRv0YiejBogZ-rut_UXVoqgkzOr6wVLMNHI']",
  "event_id=$4cZ4mezAPRv0YiejBogZ-rut_UXVoqgkzOr6wVLMNHI,depth=10,body=Historical 22 (batch=0),prevs=['$4a579_DVRTGUvvm1M5kFfer5s38Z4Z6E0aF4pjLW2y0']",
  "event_id=$4a579_DVRTGUvvm1M5kFfer5s38Z4Z6E0aF4pjLW2y0,depth=10,body=Historical 21 (batch=0),prevs=['$FqkQwqOeYGb1Wlifrft0cy_2ku_xLZIfM2VVOk9M9VM']",
  "event_id=$FqkQwqOeYGb1Wlifrft0cy_2ku_xLZIfM2VVOk9M9VM,depth=10,body=Historical 20 (batch=0),prevs=['$RcB9X6Ak1RzFX8bi_yYRaxKBKLAOYzCwSU6Ig0EZ3xs']",
  "event_id=$RcB9X6Ak1RzFX8bi_yYRaxKBKLAOYzCwSU6Ig0EZ3xs,depth=10,body=Historical 19 (batch=0),prevs=['$c86BfxVfeDAxTXVRMN3cwNe53VhjqpJnjckEbIqYuh8']",
  "event_id=$c86BfxVfeDAxTXVRMN3cwNe53VhjqpJnjckEbIqYuh8,depth=10,body=Historical 18 (batch=0),prevs=['$mZBBoO5x7B3MqEVX0gkD8QC7F9YxfC0GhWwjZhshvy8']",
  "event_id=$mZBBoO5x7B3MqEVX0gkD8QC7F9YxfC0GhWwjZhshvy8,depth=10,body=Historical 17 (batch=0),prevs=['$QKA62c3KVA_BP6TcFxDufXXZdEtpyij0eLdgg5q5jH4']",
  "event_id=$QKA62c3KVA_BP6TcFxDufXXZdEtpyij0eLdgg5q5jH4,depth=10,body=Historical 16 (batch=0),prevs=['$vjq6iv-E23ZCTSnVBbZ7ITQDnx7xyBuDQs6PJ1m88MM']",
  "event_id=$vjq6iv-E23ZCTSnVBbZ7ITQDnx7xyBuDQs6PJ1m88MM,depth=10,body=Historical 15 (batch=0),prevs=['$FlXcl0iDAnor__tQvX_TJ0SyUQQ1pLC6Cj120yaLEIA']",
  "event_id=$FlXcl0iDAnor__tQvX_TJ0SyUQQ1pLC6Cj120yaLEIA,depth=10,body=Historical 14 (batch=0),prevs=['$FN86rh1bGNpgR7YSvoov-CY6N1uQWer_1Q_agMgI-PY']",
  "event_id=$FN86rh1bGNpgR7YSvoov-CY6N1uQWer_1Q_agMgI-PY,depth=10,body=Historical 13 (batch=0),prevs=['$glaaPx9K_HNHCjKJLnWI9o9_lZ2ixy8rGVt7plkoG2M']",
  "event_id=$glaaPx9K_HNHCjKJLnWI9o9_lZ2ixy8rGVt7plkoG2M,depth=10,body=Historical 12 (batch=0),prevs=['$TrNNaavKsAVVVoxI-Dx1B6HyvGCCqHBFRJyeaOR3yEE']",
  "event_id=$TrNNaavKsAVVVoxI-Dx1B6HyvGCCqHBFRJyeaOR3yEE,depth=10,body=Historical 11 (batch=0),prevs=['$kqhpCqT74SIw33FGha6WtfTzxldtkmAGdvM0oDzsrro']",
  "event_id=$kqhpCqT74SIw33FGha6WtfTzxldtkmAGdvM0oDzsrro,depth=10,body=Historical 10 (batch=0),prevs=['$JW9bE96eX7qYwZVt_QGrSSNf7iYzMLlgJUPHBTk1AYE']",
  "event_id=$JW9bE96eX7qYwZVt_QGrSSNf7iYzMLlgJUPHBTk1AYE,depth=10,body=Historical 9 (batch=0),prevs=['$uSavs7aBJCI0eQYQh-YBbtOiR7i25KJxWkY0b1CaewE']",
  "event_id=$uSavs7aBJCI0eQYQh-YBbtOiR7i25KJxWkY0b1CaewE,depth=10,body=Historical 8 (batch=0),prevs=['$Ddld5KzIB27E2SsxZrgaCAK5tCTaeVMhmVSJDSw3hkI']",
  "event_id=$Ddld5KzIB27E2SsxZrgaCAK5tCTaeVMhmVSJDSw3hkI,depth=10,body=Historical 7 (batch=0),prevs=['$prHZpuk-6KtXwh00hJ9Yl1uP07Tkp_1WahX0Wj71S_I']",
  "event_id=$prHZpuk-6KtXwh00hJ9Yl1uP07Tkp_1WahX0Wj71S_I,depth=10,body=Historical 6 (batch=0),prevs=['$sUf-beyNZ6rHSB-yPE1ktaQfcfGFgIOm0PYj6ioivOI']",
  "event_id=$sUf-beyNZ6rHSB-yPE1ktaQfcfGFgIOm0PYj6ioivOI,depth=10,body=Historical 5 (batch=0),prevs=['$0IXibe2-ii8lega-78qYUKj6g479fuY1W4QrBv39-_w']",
  "event_id=$IpMnxei7CKhb5VgpqSpV3WRwvH8BTkxTUd8QudPyj64,depth=11,body=Message 1 (eventIDsAfter),prevs=['$iB7TihEgIcj-GSx61JLgkD7SPJXOvdpgRheXqIhCCjk']",
  "event_id=$iB7TihEgIcj-GSx61JLgkD7SPJXOvdpgRheXqIhCCjk,depth=10,body=Message 0 (eventIDsAfter),prevs=['$bXkvCPq7wkCIqEP6x9vyhI-s2oD8HdS4XFVOFvt_w8o']",
  "event_id=$QRjoGJoIiL--s5qHAj2tKBLud1XH5WrAaI5keO6c2TQ,depth=10,body=org.matrix.msc2716.insertion,prevs=['$bXkvCPq7wkCIqEP6x9vyhI-s2oD8HdS4XFVOFvt_w8o']",
  "event_id=$0IXibe2-ii8lega-78qYUKj6g479fuY1W4QrBv39-_w,depth=10,body=Historical 4 (batch=0),prevs=['$k0rygE84PVS1WpdOYmX3TdyvkDjwBCuL8PtQqyS9mzM']"
]
```




#### Add a fake edge connection `insertion` -> `batch`

Now let's see what happens when we add fake edges that connect `insertion` events to `batch` events.

`eventIDsAfter`(B) still gets sorted in the middle of the historical messages because `Message 0 (eventIDsAfter)` and the insertion event both have the same `prev_event`(A) so it doesn't know which one should come first.

[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBCIC0tLS0tLS0tLS0tLS0tLS0tPiBBXG4gICAgZW5kXG4gICAgXG4gICAgc3ViZ3JhcGggYmF0Y2gwXG4gICAgICAgIGJhdGNoMC1iYXRjaFtbXCJiYXRjaFwiXV0gLS0-IGJhdGNoMC01KChcIjVcIikpIC0tPiBiYXRjaDAtNCgoXCI0XCIpKSAtLT4gYmF0Y2gwLTMoKFwiM1wiKSkgLS0-IGJhY2tmaWxsRW5kKFwidGhpcyBpcyBhcyBmYXIgYXMgYmFja2ZpbGwgcmVhY2hlZCAoaW1hZ2luZSAxMDAgZXZlbnRzKVwiKVxuICAgIGVuZFxuXG5cbiAgICBiYXRjaDAtaW5zZXJ0aW9uQmFzZSAtLSBmYWtlIGVkZ2UgLS0-IGJhdGNoMC1iYXRjaCBcblxuICAgIGJhdGNoMC1pbnNlcnRpb25CYXNlWy9pbnNlcnRpb25cXF0gLS0tLS0tLS0tLS0tLS0tPiBBXG4iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)
![](https://user-images.githubusercontent.com/558581/139520268-e611bc3b-370e-43a1-8f26-0d2eb215b31b.png)

```
backfill sorted_events=[
  "event_id=$UmkHjkf3_9l-fwUdJrsikRIuJhXXAyXh9-pLvsY4vB0,depth=10,body=org.matrix.msc2716.insertion,prevs=['$dffYCZQmucv2wlv_dyE6-Y3bkB2I51fR4eHb2WNSdfs']",
  "event_id=$f0ohujib7hgzL8g8LHy7PL1_ONxZnnfX8AOyDp2aD3U,depth=10,body=org.matrix.msc2716.batch,prevs=['$M3bxG9qozqYi8piELdCFtx7OU1dlzb613e5CxERm8y0']",
  "event_id=$M3bxG9qozqYi8piELdCFtx7OU1dlzb613e5CxERm8y0,depth=10,body=Historical 99 (batch=0),prevs=['$KYpmMNltgP_9Qp-x3dEz63X5nNubor86s8PtI-7bR-M']",
  "event_id=$KYpmMNltgP_9Qp-x3dEz63X5nNubor86s8PtI-7bR-M,depth=10,body=Historical 98 (batch=0),prevs=['$LL2FHy9DMJJAuTIv1SR5A9eSmZIgH399C8vxh2GRzSc']",
  "event_id=$LL2FHy9DMJJAuTIv1SR5A9eSmZIgH399C8vxh2GRzSc,depth=10,body=Historical 97 (batch=0),prevs=['$fXTZDCEpgVLJMKCruH356Qna5OHARkEfZki9Vm88VSE']",
  "event_id=$fXTZDCEpgVLJMKCruH356Qna5OHARkEfZki9Vm88VSE,depth=10,body=Historical 96 (batch=0),prevs=['$9DLgWxX6c6uxK803F7RcHebwF87hpjuSf9-jbts1p8I']",
  "event_id=$9DLgWxX6c6uxK803F7RcHebwF87hpjuSf9-jbts1p8I,depth=10,body=Historical 95 (batch=0),prevs=['$NSTKo7rKQrdRsi6wo0DYGqmNn9W-EjGQ6mtvvcdMLY8']",
  "event_id=$NSTKo7rKQrdRsi6wo0DYGqmNn9W-EjGQ6mtvvcdMLY8,depth=10,body=Historical 94 (batch=0),prevs=['$kjBRg2ICcOIP4NPoa8UL3-bHqegkDr2HDp213pxL-68']",
  "event_id=$kjBRg2ICcOIP4NPoa8UL3-bHqegkDr2HDp213pxL-68,depth=10,body=Historical 93 (batch=0),prevs=['$CUjGQVg7Uy8q_wXBqW7XYXi5dBzWBJLDjrwoma3N7Ic']",
  "event_id=$CUjGQVg7Uy8q_wXBqW7XYXi5dBzWBJLDjrwoma3N7Ic,depth=10,body=Historical 92 (batch=0),prevs=['$igLomjG6VAkmzUJrn2vICSZ6yGWd3nrt75VUDwcUrao']",
  "event_id=$igLomjG6VAkmzUJrn2vICSZ6yGWd3nrt75VUDwcUrao,depth=10,body=Historical 91 (batch=0),prevs=['$LGvrmNrw8e_QkVdN6lNZshh-wVReviaguRrZqbmYYVE']",
  "event_id=$LGvrmNrw8e_QkVdN6lNZshh-wVReviaguRrZqbmYYVE,depth=10,body=Historical 90 (batch=0),prevs=['$yFotSHoAE4IUixwhZTQ8S8qIwFvZNwoYp3Z8W0VrspY']",
  "event_id=$yFotSHoAE4IUixwhZTQ8S8qIwFvZNwoYp3Z8W0VrspY,depth=10,body=Historical 89 (batch=0),prevs=['$nzx9bO52tbkbwVfb46izIA37SrWq7apic3GlwZpcM6Q']",
  "event_id=$nzx9bO52tbkbwVfb46izIA37SrWq7apic3GlwZpcM6Q,depth=10,body=Historical 88 (batch=0),prevs=['$KnnuZsNfIS4LqKtYWstzEBSOTRvetUu9sZbNDE-FXs0']",
  "event_id=$KnnuZsNfIS4LqKtYWstzEBSOTRvetUu9sZbNDE-FXs0,depth=10,body=Historical 87 (batch=0),prevs=['$LVLG1rUSf3dAMM5qb3jbC_crIQhD05E6-rF2lR4uCwQ']",
  "event_id=$LVLG1rUSf3dAMM5qb3jbC_crIQhD05E6-rF2lR4uCwQ,depth=10,body=Historical 86 (batch=0),prevs=['$U-krRAzhL75pcT_eWBSvuHftsURPkBK3eY98W6u_QzQ']",
  "event_id=$U-krRAzhL75pcT_eWBSvuHftsURPkBK3eY98W6u_QzQ,depth=10,body=Historical 85 (batch=0),prevs=['$cyc48LrVm0-igeZemrvFjW64-JKj2m6jCq2fsPpLoyI']",
  "event_id=$cyc48LrVm0-igeZemrvFjW64-JKj2m6jCq2fsPpLoyI,depth=10,body=Historical 84 (batch=0),prevs=['$Wg0V2h4QhY3SRpAYfNZCmv4HiCdEWowheIp6V8jrZqU']",
  "event_id=$Wg0V2h4QhY3SRpAYfNZCmv4HiCdEWowheIp6V8jrZqU,depth=10,body=Historical 83 (batch=0),prevs=['$PMKlvU_z_KjnWrWsFNWzR6p3ZYKQTuyFnf_bgdsGtXA']",
  "event_id=$PMKlvU_z_KjnWrWsFNWzR6p3ZYKQTuyFnf_bgdsGtXA,depth=10,body=Historical 82 (batch=0),prevs=['$7g9BS7P434JaX9VW7F2d10vsAbGMsYpRGTuWxOPvtOg']",
  "event_id=$7g9BS7P434JaX9VW7F2d10vsAbGMsYpRGTuWxOPvtOg,depth=10,body=Historical 81 (batch=0),prevs=['$1aVFrQ4wZh7ajdiM-wV3Kfk_jGIzn446pYXsZJofBkA']",
  "event_id=$1aVFrQ4wZh7ajdiM-wV3Kfk_jGIzn446pYXsZJofBkA,depth=10,body=Historical 80 (batch=0),prevs=['$Hm5CdgVY8Gz3CqFQGe_Ssqi7lxuxWkrnsRJqx2HLKwU']",
  "event_id=$Hm5CdgVY8Gz3CqFQGe_Ssqi7lxuxWkrnsRJqx2HLKwU,depth=10,body=Historical 79 (batch=0),prevs=['$tTQMkrrKjhmm_8DyeKtHBBxN5BGDtW8W136i4k8Ctis']",
  "event_id=$tTQMkrrKjhmm_8DyeKtHBBxN5BGDtW8W136i4k8Ctis,depth=10,body=Historical 78 (batch=0),prevs=['$1qPGBQbC-KJbE3h87s4ZNO5C3bRVD0zlrUIK2Olggf4']",
  "event_id=$1qPGBQbC-KJbE3h87s4ZNO5C3bRVD0zlrUIK2Olggf4,depth=10,body=Historical 77 (batch=0),prevs=['$_U98S9dfY-e6jS3nlacXFg1tXHpHWhvtb2UyNMQU3C0']",
  "event_id=$_U98S9dfY-e6jS3nlacXFg1tXHpHWhvtb2UyNMQU3C0,depth=10,body=Historical 76 (batch=0),prevs=['$NE1QLw3ybyRKrRfZTxBZvNtAugxN6dqYc36AYtHRnAg']",
  "event_id=$NE1QLw3ybyRKrRfZTxBZvNtAugxN6dqYc36AYtHRnAg,depth=10,body=Historical 75 (batch=0),prevs=['$KYRNIpHndR7Ulpna6F-XhmXJK-eLYXPiOGUS0e-mxR0']",
  "event_id=$KYRNIpHndR7Ulpna6F-XhmXJK-eLYXPiOGUS0e-mxR0,depth=10,body=Historical 74 (batch=0),prevs=['$kiCMmtu6EMEYJOVYHC75-I8ZrMyItRliS7XQ4exYMKk']",
  "event_id=$kiCMmtu6EMEYJOVYHC75-I8ZrMyItRliS7XQ4exYMKk,depth=10,body=Historical 73 (batch=0),prevs=['$RULeZdv9MA3KTxiyvJC3l_3kWAy6dRiLX-6uLyUu9cU']",
  "event_id=$RULeZdv9MA3KTxiyvJC3l_3kWAy6dRiLX-6uLyUu9cU,depth=10,body=Historical 72 (batch=0),prevs=['$N_dKO_GxgGEKaB5o8pMnb18H3SyY85mrX-szGutZ79Q']",
  "event_id=$N_dKO_GxgGEKaB5o8pMnb18H3SyY85mrX-szGutZ79Q,depth=10,body=Historical 71 (batch=0),prevs=['$suS0_3L0XJpFHpQRokRIBOcHFiiK-OKmZ1aeTRQA2yM']",
  "event_id=$suS0_3L0XJpFHpQRokRIBOcHFiiK-OKmZ1aeTRQA2yM,depth=10,body=Historical 70 (batch=0),prevs=['$qLBZ2xN0lHbQ2PdFlwL45e5Wxb-JKRwiQu8Eqekh918']",
  "event_id=$qLBZ2xN0lHbQ2PdFlwL45e5Wxb-JKRwiQu8Eqekh918,depth=10,body=Historical 69 (batch=0),prevs=['$-T5Ii2ddpnJ-sRyi5rbadeFViyE_PWvHVoFEcV4eqlU']",
  "event_id=$-T5Ii2ddpnJ-sRyi5rbadeFViyE_PWvHVoFEcV4eqlU,depth=10,body=Historical 68 (batch=0),prevs=['$JAL6R940NjRippf3v1ubxDlxMh3RKCYn_t-uLzLCiUk']",
  "event_id=$JAL6R940NjRippf3v1ubxDlxMh3RKCYn_t-uLzLCiUk,depth=10,body=Historical 67 (batch=0),prevs=['$kvySNMcyak-FzEJuEHZ7281omsl9ON6OZA9MfI0r_m0']",
  "event_id=$kvySNMcyak-FzEJuEHZ7281omsl9ON6OZA9MfI0r_m0,depth=10,body=Historical 66 (batch=0),prevs=['$xU6BU-5ygwMrpOPQKULjCRUz457lZ_020fSryvDx2mM']",
  "event_id=$xU6BU-5ygwMrpOPQKULjCRUz457lZ_020fSryvDx2mM,depth=10,body=Historical 65 (batch=0),prevs=['$aFXDqrp2XXIH06jgVi5WmjF6gzzsKxuaWGE0Zu9cgmU']",
  "event_id=$aFXDqrp2XXIH06jgVi5WmjF6gzzsKxuaWGE0Zu9cgmU,depth=10,body=Historical 64 (batch=0),prevs=['$ulSikUEZZSNZ9iiNud7eS662v6wJtCfAelXUWkHP8t4']",
  "event_id=$ulSikUEZZSNZ9iiNud7eS662v6wJtCfAelXUWkHP8t4,depth=10,body=Historical 63 (batch=0),prevs=['$JRhI_Ey2zjnpVCWfyywUFfycSfyHeVWXvPEvxGqLeAk']",
  "event_id=$JRhI_Ey2zjnpVCWfyywUFfycSfyHeVWXvPEvxGqLeAk,depth=10,body=Historical 62 (batch=0),prevs=['$8gqVZ-h27SEzAnd7uUSpO3HPUCHkPj3nSJVwj8__yAk']",
  "event_id=$8gqVZ-h27SEzAnd7uUSpO3HPUCHkPj3nSJVwj8__yAk,depth=10,body=Historical 61 (batch=0),prevs=['$-_yO9osiHl4H8A0tIjkjf109v4N-qubBeWs1aejuTyA']",
  "event_id=$-_yO9osiHl4H8A0tIjkjf109v4N-qubBeWs1aejuTyA,depth=10,body=Historical 60 (batch=0),prevs=['$9FZIfbf-C9562OD6_yj0V9-biX790KeptVoAiZr3K5s']",
  "event_id=$9FZIfbf-C9562OD6_yj0V9-biX790KeptVoAiZr3K5s,depth=10,body=Historical 59 (batch=0),prevs=['$kLPwjL2BVXuQ4MO-hqlPwXqSwlcYzV0jx_1T7xDI9D4']",
  "event_id=$kLPwjL2BVXuQ4MO-hqlPwXqSwlcYzV0jx_1T7xDI9D4,depth=10,body=Historical 58 (batch=0),prevs=['$0mB49ALssJmHwc6qL6dMOGri73ICfVE6g4v7UjjD8XE']",
  "event_id=$0mB49ALssJmHwc6qL6dMOGri73ICfVE6g4v7UjjD8XE,depth=10,body=Historical 57 (batch=0),prevs=['$4Iw1zPXkDGUULkfVRifjTbKoW5D85pVIkorxD66DHMo']",
  "event_id=$4Iw1zPXkDGUULkfVRifjTbKoW5D85pVIkorxD66DHMo,depth=10,body=Historical 56 (batch=0),prevs=['$0WXpCQCsNWzd2vBHCrYiVFByhWbpnaxfYREPNowrcic']",
  "event_id=$0WXpCQCsNWzd2vBHCrYiVFByhWbpnaxfYREPNowrcic,depth=10,body=Historical 55 (batch=0),prevs=['$ZF4ufRI_wJu4Rbc9iHiLOVLsQ1K1d9keieysJ2aj3Ec']",
  "event_id=$ZF4ufRI_wJu4Rbc9iHiLOVLsQ1K1d9keieysJ2aj3Ec,depth=10,body=Historical 54 (batch=0),prevs=['$UKa2BcuwQX_WlHCj_GP4ZT15Uf4A_oAMBHGi8SzAv48']",
  "event_id=$UKa2BcuwQX_WlHCj_GP4ZT15Uf4A_oAMBHGi8SzAv48,depth=10,body=Historical 53 (batch=0),prevs=['$f2HPlE68rwAEGrz87J9NdiaFi1b3xzgF3PlcCoRNBTo']",
  "event_id=$f2HPlE68rwAEGrz87J9NdiaFi1b3xzgF3PlcCoRNBTo,depth=10,body=Historical 52 (batch=0),prevs=['$qEoC_6YMmEYaPX6Zx34EPWFi7qBtHtmoD2qNGkJrN0A']",
  "event_id=$qEoC_6YMmEYaPX6Zx34EPWFi7qBtHtmoD2qNGkJrN0A,depth=10,body=Historical 51 (batch=0),prevs=['$N8p0kJwuhvX3lWJvpaBxX5x7DAlak8Y42A96cwgQ4uw']",
  "event_id=$N8p0kJwuhvX3lWJvpaBxX5x7DAlak8Y42A96cwgQ4uw,depth=10,body=Historical 50 (batch=0),prevs=['$9rKVaFmQFyK3ySwPAOEY0lbb_9ZCTWry_LmMLNNooKM']",
  "event_id=$9rKVaFmQFyK3ySwPAOEY0lbb_9ZCTWry_LmMLNNooKM,depth=10,body=Historical 49 (batch=0),prevs=['$bCm8Bu1NgcgHEuvN4qWnVArIYjPjxmvIMNft2UUjWCw']",
  "event_id=$bCm8Bu1NgcgHEuvN4qWnVArIYjPjxmvIMNft2UUjWCw,depth=10,body=Historical 48 (batch=0),prevs=['$zePm6HscluBTmWr8SKk2Cr2fQHTjYSOU44002uHmAJo']",
  "event_id=$zePm6HscluBTmWr8SKk2Cr2fQHTjYSOU44002uHmAJo,depth=10,body=Historical 47 (batch=0),prevs=['$xqR3vY91H2ZWg65qqnaF3JvEj22oeUFrdmKmTOUX81M']",
  "event_id=$xqR3vY91H2ZWg65qqnaF3JvEj22oeUFrdmKmTOUX81M,depth=10,body=Historical 46 (batch=0),prevs=['$i7iZZJ648LpquLUAZUXNCRQu0eTfejMDCtFLZY-OB9Q']",
  "event_id=$i7iZZJ648LpquLUAZUXNCRQu0eTfejMDCtFLZY-OB9Q,depth=10,body=Historical 45 (batch=0),prevs=['$tx0qtZa8yoSHvg_0kKJrXUvueNLDIeNuKw3lqcV3Ims']",
  "event_id=$tx0qtZa8yoSHvg_0kKJrXUvueNLDIeNuKw3lqcV3Ims,depth=10,body=Historical 44 (batch=0),prevs=['$l0_-kM03G4NoxMmb6EnhiyqcYF1JDmzJHQrGeMqxUUY']",
  "event_id=$l0_-kM03G4NoxMmb6EnhiyqcYF1JDmzJHQrGeMqxUUY,depth=10,body=Historical 43 (batch=0),prevs=['$F11qmdJOjACj-3EiNuNqqE7LWQcfWw2-GMejrUC_258']",
  "event_id=$F11qmdJOjACj-3EiNuNqqE7LWQcfWw2-GMejrUC_258,depth=10,body=Historical 42 (batch=0),prevs=['$SHZn93xx6uaD3CLnjCyCwGS9ic_9MdwVd0EwA87Ygdo']",
  "event_id=$SHZn93xx6uaD3CLnjCyCwGS9ic_9MdwVd0EwA87Ygdo,depth=10,body=Historical 41 (batch=0),prevs=['$ClqKnQwyPlIe_943ktsQT725sLlnSWWUpDhds_uQCZA']",
  "event_id=$ClqKnQwyPlIe_943ktsQT725sLlnSWWUpDhds_uQCZA,depth=10,body=Historical 40 (batch=0),prevs=['$2RqtwO8JggYVR0CZ_A623ciBQUkN3gPg8XXrICEvIA8']",
  "event_id=$2RqtwO8JggYVR0CZ_A623ciBQUkN3gPg8XXrICEvIA8,depth=10,body=Historical 39 (batch=0),prevs=['$9c4EEDCYchVckxL4I4dm7i5zNzgB9nQ9CgTgGAmOU80']",
  "event_id=$9c4EEDCYchVckxL4I4dm7i5zNzgB9nQ9CgTgGAmOU80,depth=10,body=Historical 38 (batch=0),prevs=['$jpUITAKRARGRABZAGvZVSctl7ZCXt-n_xRY8sEehlzg']",
  "event_id=$jpUITAKRARGRABZAGvZVSctl7ZCXt-n_xRY8sEehlzg,depth=10,body=Historical 37 (batch=0),prevs=['$XdtyLIWT6BdP41cbP3Cu-EM9kqzRllURmRy8FTaUfsk']",
  "event_id=$XdtyLIWT6BdP41cbP3Cu-EM9kqzRllURmRy8FTaUfsk,depth=10,body=Historical 36 (batch=0),prevs=['$S1SFz7aIzhnXiwtbbVYTbCiyJepFQm1E8q2mC_t19uU']",
  "event_id=$S1SFz7aIzhnXiwtbbVYTbCiyJepFQm1E8q2mC_t19uU,depth=10,body=Historical 35 (batch=0),prevs=['$LBTQunwZfaYEJR-Z92B0qItTiLLcDzX824Au7xkpph0']",
  "event_id=$LBTQunwZfaYEJR-Z92B0qItTiLLcDzX824Au7xkpph0,depth=10,body=Historical 34 (batch=0),prevs=['$gYdeBoOkFBXHUV9ZpW9_vLoZeQTVqEhZ-7s-YW0H1VA']",
  "event_id=$gYdeBoOkFBXHUV9ZpW9_vLoZeQTVqEhZ-7s-YW0H1VA,depth=10,body=Historical 33 (batch=0),prevs=['$MX_d_wvZrv97162jipUbLPs9Ce1MxDA-zKJDozKSzgQ']",
  "event_id=$MX_d_wvZrv97162jipUbLPs9Ce1MxDA-zKJDozKSzgQ,depth=10,body=Historical 32 (batch=0),prevs=['$WjYDkrAUt-EZi4M6OLvot5jWgqJCaPGEO1ivf_F-Skw']",
  "event_id=$WjYDkrAUt-EZi4M6OLvot5jWgqJCaPGEO1ivf_F-Skw,depth=10,body=Historical 31 (batch=0),prevs=['$I44xv5Lb1v0eABbaNKHxAdOcbBpgyjfmSqsps1TvjXw']",
  "event_id=$I44xv5Lb1v0eABbaNKHxAdOcbBpgyjfmSqsps1TvjXw,depth=10,body=Historical 30 (batch=0),prevs=['$xtPtcjWXcm84gJkwoZlgQn8ljmgM7pc-1plFKxrFNx0']",
  "event_id=$xtPtcjWXcm84gJkwoZlgQn8ljmgM7pc-1plFKxrFNx0,depth=10,body=Historical 29 (batch=0),prevs=['$LOP8Bs7wiCYEgc4YweByPfnEhcOcl-gsfZkTl1Y1TZ4']",
  "event_id=$LOP8Bs7wiCYEgc4YweByPfnEhcOcl-gsfZkTl1Y1TZ4,depth=10,body=Historical 28 (batch=0),prevs=['$wSD475pGliwMDj3x8ZeIsuTdFYcuJOUQkcVwZw0VlE4']",
  "event_id=$wSD475pGliwMDj3x8ZeIsuTdFYcuJOUQkcVwZw0VlE4,depth=10,body=Historical 27 (batch=0),prevs=['$iMBEDY81GuL1xaRM4lRu6uQF3y17d60zVv1IKB7Rlkw']",
  "event_id=$cz-wA9MtNkqJmmkqDbo9EnVkpBv7khBlB6-i1UjGMS0,depth=11,body=Message 1 (eventIDsAfter),prevs=['$vHYRLWg3-0sCad3g42qInThwi2_6-3x-iGawDgBTN6s']",
  "event_id=$vHYRLWg3-0sCad3g42qInThwi2_6-3x-iGawDgBTN6s,depth=10,body=Message 0 (eventIDsAfter),prevs=['$dffYCZQmucv2wlv_dyE6-Y3bkB2I51fR4eHb2WNSdfs']",
  "event_id=$iMBEDY81GuL1xaRM4lRu6uQF3y17d60zVv1IKB7Rlkw,depth=10,body=Historical 26 (batch=0),prevs=['$PUqqPFb-v8j-dYw1QCMKLI6xf1LrIrDXlxNkO26JdZw']",
  "event_id=$PUqqPFb-v8j-dYw1QCMKLI6xf1LrIrDXlxNkO26JdZw,depth=10,body=Historical 25 (batch=0),prevs=['$2ikbYbJ4x3eyrxTX8RZD8cX9WmLGcZRgIU2CiIl_rE0']",
  "event_id=$2ikbYbJ4x3eyrxTX8RZD8cX9WmLGcZRgIU2CiIl_rE0,depth=10,body=Historical 24 (batch=0),prevs=['$CavjQ3MsYjDo0O5rcfOsnRxwQlXpWyil3l_pNOkh1hg']",
  "event_id=$CavjQ3MsYjDo0O5rcfOsnRxwQlXpWyil3l_pNOkh1hg,depth=10,body=Historical 23 (batch=0),prevs=['$qMWS3vZFQQq_nrYGbLWN84R0Abd9mkrHSD007TvNCs0']",
  "event_id=$qMWS3vZFQQq_nrYGbLWN84R0Abd9mkrHSD007TvNCs0,depth=10,body=Historical 22 (batch=0),prevs=['$kQWUARyeFZFKyyLaAdOX6rRWIOm2luJIu9CW3qsWX5s']",
  "event_id=$kQWUARyeFZFKyyLaAdOX6rRWIOm2luJIu9CW3qsWX5s,depth=10,body=Historical 21 (batch=0),prevs=['$dEzSF44x22bgpprgZb2YdM0yAPim4QuuuqY5kte-JTE']",
  "event_id=$dEzSF44x22bgpprgZb2YdM0yAPim4QuuuqY5kte-JTE,depth=10,body=Historical 20 (batch=0),prevs=['$74LMlNOIuTIlHadYg9KkpqMjB2VfVj4qbekOoZo_Ecs']",
  "event_id=$74LMlNOIuTIlHadYg9KkpqMjB2VfVj4qbekOoZo_Ecs,depth=10,body=Historical 19 (batch=0),prevs=['$-hvXG5ufi-chHVsK4SEbL54dmra34SgEXLkMIdTaFFI']",
  "event_id=$-hvXG5ufi-chHVsK4SEbL54dmra34SgEXLkMIdTaFFI,depth=10,body=Historical 18 (batch=0),prevs=['$8dpEKuhTeJYnbexsgNRD3qUgMBWqNkMOynp4o7D9-Cw']",
  "event_id=$8dpEKuhTeJYnbexsgNRD3qUgMBWqNkMOynp4o7D9-Cw,depth=10,body=Historical 17 (batch=0),prevs=['$1hYt8JZTpT5q0C3y_yZ1n7TQCc48NRgOSMcUSNvvap4']",
  "event_id=$1hYt8JZTpT5q0C3y_yZ1n7TQCc48NRgOSMcUSNvvap4,depth=10,body=Historical 16 (batch=0),prevs=['$l8XF6K8UEttTfdzP8ONM23KhpeS9k1ngpbtDD3jbBYg']",
  "event_id=$l8XF6K8UEttTfdzP8ONM23KhpeS9k1ngpbtDD3jbBYg,depth=10,body=Historical 15 (batch=0),prevs=['$ExS8bmzBXJC9RDwyIFdWWbTuRBwExXzSCBN9l5qxJZo']",
  "event_id=$ExS8bmzBXJC9RDwyIFdWWbTuRBwExXzSCBN9l5qxJZo,depth=10,body=Historical 14 (batch=0),prevs=['$IKxIEbqaQaiVS3X72LbUxHkqS-xgvs5eZ5e6gRtqWa8']",
  "event_id=$IKxIEbqaQaiVS3X72LbUxHkqS-xgvs5eZ5e6gRtqWa8,depth=10,body=Historical 13 (batch=0),prevs=['$04IRMqUUh4Ox9VZmuCApeWQ0Vles-_iLtpzeaSSN_Ps']",
  "event_id=$04IRMqUUh4Ox9VZmuCApeWQ0Vles-_iLtpzeaSSN_Ps,depth=10,body=Historical 12 (batch=0),prevs=['$7qrzXJLLdFQ_BuZlDkYnlpeBJE5xMkZEC-qAyfSVl00']",
  "event_id=$7qrzXJLLdFQ_BuZlDkYnlpeBJE5xMkZEC-qAyfSVl00,depth=10,body=Historical 11 (batch=0),prevs=['$Bk63Ka0tDDusycrIG9zrWA8E1n9xoFbm99qJ0kfZwyc']",
  "event_id=$Bk63Ka0tDDusycrIG9zrWA8E1n9xoFbm99qJ0kfZwyc,depth=10,body=Historical 10 (batch=0),prevs=['$T3LksRSD-nKVONC1cRRZgArbqIzexyntGfC44e3BUsw']",
  "event_id=$T3LksRSD-nKVONC1cRRZgArbqIzexyntGfC44e3BUsw,depth=10,body=Historical 9 (batch=0),prevs=['$raczLD9wYYgm03LIsnRSeIAadqIqs5lWH6hHkDRxd1g']",
  "event_id=$raczLD9wYYgm03LIsnRSeIAadqIqs5lWH6hHkDRxd1g,depth=10,body=Historical 8 (batch=0),prevs=['$A5F-OG6Dl4IQWkAhQPxbqXCYEQBKASUvwoVStKBFoWw']",
  "event_id=$A5F-OG6Dl4IQWkAhQPxbqXCYEQBKASUvwoVStKBFoWw,depth=10,body=Historical 7 (batch=0),prevs=['$CuBJxdfQMLHO72cgjjb3nrVEQX7_eMzIujdKp5BSs4I']",
  "event_id=$CuBJxdfQMLHO72cgjjb3nrVEQX7_eMzIujdKp5BSs4I,depth=10,body=Historical 6 (batch=0),prevs=['$1zFLe4nZZoisGt4RTJO-gq-n_f69ruMxgTTbv3CuDLI']",
  "event_id=$1zFLe4nZZoisGt4RTJO-gq-n_f69ruMxgTTbv3CuDLI,depth=10,body=Historical 5 (batch=0),prevs=['$QgVfQHOCKQWPdupDl4qQvMYadzgGgvoRoITFqZTr7Qk']",
  "event_id=$QgVfQHOCKQWPdupDl4qQvMYadzgGgvoRoITFqZTr7Qk,depth=10,body=Historical 4 (batch=0),prevs=['$dJzjjiQ-1wCY4dJkoLC6fvs70Hi2gpxEs4eo2TRhZO0']"
]
```



#### Add fake edges connecting `insertion` -> `batch` and the insertion events `prev_event` successors -> insertion event ✅

 - `insertion` -> `batch`
 - The insertion events `prev_event` successors -> insertion event
 
 This works great to solve all of the problems when we have `eventIDsAfter`(B) and part of historical batch in the same backfill chunk. Now it knows `eventIDsAfter`(B) comes after the historical batch. And matches the expected sort ✅
 
 
[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBCIC0tLS0tLS0tLS0tLS0tLS0tPiBBXG4gICAgZW5kXG4gICAgXG4gICAgc3ViZ3JhcGggYmF0Y2gwXG4gICAgICAgIGJhdGNoMC1iYXRjaFtbXCJiYXRjaFwiXV0gLS0-IGJhdGNoMC01KChcIjVcIikpIC0tPiBiYXRjaDAtNCgoXCI0XCIpKSAtLT4gYmF0Y2gwLTMoKFwiM1wiKSkgLS0-IGJhY2tmaWxsRW5kKFwidGhpcyBpcyBhcyBmYXIgYXMgYmFja2ZpbGwgcmVhY2hlZCAoaW1hZ2luZSAxMDAgZXZlbnRzKVwiKVxuICAgIGVuZFxuXG5cbiAgICBiYXRjaDAtaW5zZXJ0aW9uQmFzZSAtLSBmYWtlIGVkZ2UgLS0-IGJhdGNoMC1iYXRjaCBcbiAgICBCIC0tIGZha2UgZWRnZSAtLT4gYmF0Y2gwLWluc2VydGlvbkJhc2VcblxuICAgIGJhdGNoMC1pbnNlcnRpb25CYXNlWy9pbnNlcnRpb25cXF0gLS0tLS0tLS0tLS0tLS0tPiBBXG4iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)
![](https://user-images.githubusercontent.com/558581/139520695-9214e63a-a8d0-4330-9924-0b19bea6f4a3.png)


 ```
 backfill sorted_events=[
  "event_id=$c314gfzp5eK3Ti7CFADPe5DV31ew4P-Sam_tLd70QB4,depth=11,body=Message 1 (eventIDsAfter),prevs=['$Tw6QXdermFoW4JQcc7Zu5U3Khlz1qVG6BD6Vq6VmiUY']",
  "event_id=$Tw6QXdermFoW4JQcc7Zu5U3Khlz1qVG6BD6Vq6VmiUY,depth=10,body=Message 0 (eventIDsAfter),prevs=['$PVJJWg-rYE1a3TLEAQG-1HexD-b6QAFnlgJjdJooF8Y']",
  "event_id=$_ktfy1Mzf3HF2KtabjoK48pDiBcJ2CjUWya6DagKxvg,depth=10,body=org.matrix.msc2716.insertion,prevs=['$PVJJWg-rYE1a3TLEAQG-1HexD-b6QAFnlgJjdJooF8Y']",
  "event_id=$Ki9MNoPjmZe0VwseDb82USns9HpE2T99FLoB8dPaQ4U,depth=10,body=org.matrix.msc2716.batch,prevs=['$3v5SvCqrBH-jYaE9IZGuF8auc9qFU6HRx_WdZMVVFWM']",
  "event_id=$3v5SvCqrBH-jYaE9IZGuF8auc9qFU6HRx_WdZMVVFWM,depth=10,body=Historical 99 (batch=0),prevs=['$BxVwvCHVWLzrWU1XfH7E6ZQo-x783GkySh3xx462YT4']",
  "event_id=$BxVwvCHVWLzrWU1XfH7E6ZQo-x783GkySh3xx462YT4,depth=10,body=Historical 98 (batch=0),prevs=['$yVdBlIMWDDesTVEYnfcwQ0Y6SDGB9ZwoU5P56pmhbi4']",
  "event_id=$yVdBlIMWDDesTVEYnfcwQ0Y6SDGB9ZwoU5P56pmhbi4,depth=10,body=Historical 97 (batch=0),prevs=['$Hol5KY-QBYzFcqkLNbxkYg9ZDYgj9Vo-pqOhsNb_bi4']",
  "event_id=$Hol5KY-QBYzFcqkLNbxkYg9ZDYgj9Vo-pqOhsNb_bi4,depth=10,body=Historical 96 (batch=0),prevs=['$xm6Ylbk4rTXrZaDZ31yrOwE-XparM4cyq41VlavlgIc']",
  "event_id=$xm6Ylbk4rTXrZaDZ31yrOwE-XparM4cyq41VlavlgIc,depth=10,body=Historical 95 (batch=0),prevs=['$2poYyo6Dao9Cvt5oJCvQO2RCWeYkKKwXOuPtcCLOqZI']",
  "event_id=$2poYyo6Dao9Cvt5oJCvQO2RCWeYkKKwXOuPtcCLOqZI,depth=10,body=Historical 94 (batch=0),prevs=['$019DNNmpib0T8t4_LAkmT3KNh70jyOU2iMJGOkScF_k']",
  "event_id=$019DNNmpib0T8t4_LAkmT3KNh70jyOU2iMJGOkScF_k,depth=10,body=Historical 93 (batch=0),prevs=['$ah4V0y53ApkH1K3X9uqUZnWIHMEzWob-d-PKO1E0LIY']",
  "event_id=$ah4V0y53ApkH1K3X9uqUZnWIHMEzWob-d-PKO1E0LIY,depth=10,body=Historical 92 (batch=0),prevs=['$s8WMqgb4DM93Gn5jgaDumnDF5wG2ufO7cgvuJSZbjkc']",
  "event_id=$s8WMqgb4DM93Gn5jgaDumnDF5wG2ufO7cgvuJSZbjkc,depth=10,body=Historical 91 (batch=0),prevs=['$vEFpS1Nl4sVsZ-JMtQrb-szRv3l9nCVthIPLH3aj6ZA']",
  "event_id=$vEFpS1Nl4sVsZ-JMtQrb-szRv3l9nCVthIPLH3aj6ZA,depth=10,body=Historical 90 (batch=0),prevs=['$QqhWUjaRT46E4uoZ4DVSJJM98rUrQvbaVO2_olk-2ug']",
  "event_id=$QqhWUjaRT46E4uoZ4DVSJJM98rUrQvbaVO2_olk-2ug,depth=10,body=Historical 89 (batch=0),prevs=['$9kBxYCUe1rMjdei643hz9101Qk7rlPQ0YJTp3iBI_tg']",
  "event_id=$9kBxYCUe1rMjdei643hz9101Qk7rlPQ0YJTp3iBI_tg,depth=10,body=Historical 88 (batch=0),prevs=['$Qt6qVnDowNisAGI_vpmxhKMLJ0VVBYrUZvyUSKnySnQ']",
  "event_id=$Qt6qVnDowNisAGI_vpmxhKMLJ0VVBYrUZvyUSKnySnQ,depth=10,body=Historical 87 (batch=0),prevs=['$qjNikTCNI9t33JEToLQKvfckZ_UhyR4DeQeNIJ1zdjQ']",
  "event_id=$qjNikTCNI9t33JEToLQKvfckZ_UhyR4DeQeNIJ1zdjQ,depth=10,body=Historical 86 (batch=0),prevs=['$-lRQLCGrs0dz_90aZrC1c4MTIC_2nsDrjBC4LpRVEqA']",
  "event_id=$-lRQLCGrs0dz_90aZrC1c4MTIC_2nsDrjBC4LpRVEqA,depth=10,body=Historical 85 (batch=0),prevs=['$luq70jw2plnfbgqIKH-f7zX0fNcM0yF0SNUgQfwR1_4']",
  "event_id=$luq70jw2plnfbgqIKH-f7zX0fNcM0yF0SNUgQfwR1_4,depth=10,body=Historical 84 (batch=0),prevs=['$-NxuKkBK024nE3PaVeuXS5TGNiz5aMUv75kvDgTrmK4']",
  "event_id=$-NxuKkBK024nE3PaVeuXS5TGNiz5aMUv75kvDgTrmK4,depth=10,body=Historical 83 (batch=0),prevs=['$pXqlXQw9LSXboKsJMFoeauTUAJQ8DhdFuoyyj3g2dMM']",
  "event_id=$pXqlXQw9LSXboKsJMFoeauTUAJQ8DhdFuoyyj3g2dMM,depth=10,body=Historical 82 (batch=0),prevs=['$BdwcbXFhSPuPi78PmlMun3vb4fKTR4fEErdZZOTHH1k']",
  "event_id=$BdwcbXFhSPuPi78PmlMun3vb4fKTR4fEErdZZOTHH1k,depth=10,body=Historical 81 (batch=0),prevs=['$40TRl3SFqb3ZblHuABoPDKtmDgRTBfs4krpewIfhHQg']",
  "event_id=$40TRl3SFqb3ZblHuABoPDKtmDgRTBfs4krpewIfhHQg,depth=10,body=Historical 80 (batch=0),prevs=['$Sgto0D5ABMxdsqcaXkKQToKkAE7B6kPRKDLAHur8ttA']",
  "event_id=$Sgto0D5ABMxdsqcaXkKQToKkAE7B6kPRKDLAHur8ttA,depth=10,body=Historical 79 (batch=0),prevs=['$ZKje9Q8NQdGd1TGgbAH3HibFjkeyV9SfVoBC1BTPFzQ']",
  "event_id=$ZKje9Q8NQdGd1TGgbAH3HibFjkeyV9SfVoBC1BTPFzQ,depth=10,body=Historical 78 (batch=0),prevs=['$-jh9Uyp7_sNlxG71GO95omuToYoNV0ryPffoYPH0lwQ']",
  "event_id=$-jh9Uyp7_sNlxG71GO95omuToYoNV0ryPffoYPH0lwQ,depth=10,body=Historical 77 (batch=0),prevs=['$ihhagD2WpZzeBmZ3HY9BFFVua5jMa2eIYHqdG3Gc-xI']",
  "event_id=$ihhagD2WpZzeBmZ3HY9BFFVua5jMa2eIYHqdG3Gc-xI,depth=10,body=Historical 76 (batch=0),prevs=['$hLmo3PC0qex7MQ5d8h7Z4Ze2HJyrgSQwT1xxJoUdlfk']",
  "event_id=$hLmo3PC0qex7MQ5d8h7Z4Ze2HJyrgSQwT1xxJoUdlfk,depth=10,body=Historical 75 (batch=0),prevs=['$NjVn157cod4FkJDVz2l-LLoaFFLml4k1pUIq4PuSBc8']",
  "event_id=$NjVn157cod4FkJDVz2l-LLoaFFLml4k1pUIq4PuSBc8,depth=10,body=Historical 74 (batch=0),prevs=['$_Y4Dva5xVYAuIqDrNpllvhpNWeLxpGYZUS66F2kJ_AA']",
  "event_id=$_Y4Dva5xVYAuIqDrNpllvhpNWeLxpGYZUS66F2kJ_AA,depth=10,body=Historical 73 (batch=0),prevs=['$rrtlUjtMaqsZggSNmIHvK15WUzkrGyIbFUbXDBRT1wE']",
  "event_id=$rrtlUjtMaqsZggSNmIHvK15WUzkrGyIbFUbXDBRT1wE,depth=10,body=Historical 72 (batch=0),prevs=['$DU1XngcegPIcLSJ0p-hXesfKohQi2D65WuiNpGefABY']",
  "event_id=$DU1XngcegPIcLSJ0p-hXesfKohQi2D65WuiNpGefABY,depth=10,body=Historical 71 (batch=0),prevs=['$orT3tv_2vP1smgMN2VUF4OLCyBWmbK0Jmt0k74ZiQvQ']",
  "event_id=$orT3tv_2vP1smgMN2VUF4OLCyBWmbK0Jmt0k74ZiQvQ,depth=10,body=Historical 70 (batch=0),prevs=['$TqkZLzllkXzlkJmi-gOQt7x9pd07OZYjebjwaR8Q8_Y']",
  "event_id=$TqkZLzllkXzlkJmi-gOQt7x9pd07OZYjebjwaR8Q8_Y,depth=10,body=Historical 69 (batch=0),prevs=['$PkBl3XjikUuHJN2HG0V3s_7xdSK8xWHAZR6lnXcmxtk']",
  "event_id=$PkBl3XjikUuHJN2HG0V3s_7xdSK8xWHAZR6lnXcmxtk,depth=10,body=Historical 68 (batch=0),prevs=['$vLSceQHhObd_787W7nGBr7N5fzfyjgtgSDUwsG9EUbM']",
  "event_id=$vLSceQHhObd_787W7nGBr7N5fzfyjgtgSDUwsG9EUbM,depth=10,body=Historical 67 (batch=0),prevs=['$WMOAwKpzmvUsfPA6XQpVJQSynwGEP9tGPyorQitKE9o']",
  "event_id=$WMOAwKpzmvUsfPA6XQpVJQSynwGEP9tGPyorQitKE9o,depth=10,body=Historical 66 (batch=0),prevs=['$gzai9umMkaVsgcXUvAnThyyGWMj76mx7A4h-4uvqDNM']",
  "event_id=$gzai9umMkaVsgcXUvAnThyyGWMj76mx7A4h-4uvqDNM,depth=10,body=Historical 65 (batch=0),prevs=['$n6tMJUwBvRKhskeoe6B1wo-b4bYS-fLWg1I-4iNY9nk']",
  "event_id=$n6tMJUwBvRKhskeoe6B1wo-b4bYS-fLWg1I-4iNY9nk,depth=10,body=Historical 64 (batch=0),prevs=['$UfrsjQy563In-yA3b9Kpo-dLBqYtO8feoVxO0B3_YG4']",
  "event_id=$UfrsjQy563In-yA3b9Kpo-dLBqYtO8feoVxO0B3_YG4,depth=10,body=Historical 63 (batch=0),prevs=['$W5lSrPSpJvY1lCkI7tsEG_gtov1H7MKURjfUmih79eU']",
  "event_id=$W5lSrPSpJvY1lCkI7tsEG_gtov1H7MKURjfUmih79eU,depth=10,body=Historical 62 (batch=0),prevs=['$BhGaW2JeYEbb0TOECWnH_ns5gZE4OYqQJbF7k3w6ygA']",
  "event_id=$BhGaW2JeYEbb0TOECWnH_ns5gZE4OYqQJbF7k3w6ygA,depth=10,body=Historical 61 (batch=0),prevs=['$Q0bGDDQvTB5xYJXhLtR_dFceVqGMCUMNECZyZNG70wI']",
  "event_id=$Q0bGDDQvTB5xYJXhLtR_dFceVqGMCUMNECZyZNG70wI,depth=10,body=Historical 60 (batch=0),prevs=['$SUmDzIs9hG-YB8FIy_HW_AO3MMiqnS_sFzAOA75mODc']",
  "event_id=$SUmDzIs9hG-YB8FIy_HW_AO3MMiqnS_sFzAOA75mODc,depth=10,body=Historical 59 (batch=0),prevs=['$9sLl1w2hU7tl5Rz1nSau68PrxgOhb_cINc3nwAxHBn4']",
  "event_id=$9sLl1w2hU7tl5Rz1nSau68PrxgOhb_cINc3nwAxHBn4,depth=10,body=Historical 58 (batch=0),prevs=['$_x7MurM519Q5S59Lw2B_-atXf_tSaj8qZqqUZRn6P6o']",
  "event_id=$_x7MurM519Q5S59Lw2B_-atXf_tSaj8qZqqUZRn6P6o,depth=10,body=Historical 57 (batch=0),prevs=['$N8wf3rsWbMB069WiWP5_JuBHY62YWbltnV_3fZraIgQ']",
  "event_id=$N8wf3rsWbMB069WiWP5_JuBHY62YWbltnV_3fZraIgQ,depth=10,body=Historical 56 (batch=0),prevs=['$7hTCl76JcG-MRtTGNqXnENZIXHfNnT_v3KXyFS08N4c']",
  "event_id=$7hTCl76JcG-MRtTGNqXnENZIXHfNnT_v3KXyFS08N4c,depth=10,body=Historical 55 (batch=0),prevs=['$EK8g07iJNozuoGvygr6UVwqyscce9zu5iyBPqNcLBxs']",
  "event_id=$EK8g07iJNozuoGvygr6UVwqyscce9zu5iyBPqNcLBxs,depth=10,body=Historical 54 (batch=0),prevs=['$JikjHhjhN1Actljtp18KmFQ8T1DCN9Vzpmcs3g1IPvA']",
  "event_id=$JikjHhjhN1Actljtp18KmFQ8T1DCN9Vzpmcs3g1IPvA,depth=10,body=Historical 53 (batch=0),prevs=['$3_vnUSH2MGaIvopTBp0wEfXkpcJOSZh_w-mIXv6eKJk']",
  "event_id=$3_vnUSH2MGaIvopTBp0wEfXkpcJOSZh_w-mIXv6eKJk,depth=10,body=Historical 52 (batch=0),prevs=['$7oGQwQiBEuL1XbpVzLh4ofeIWC8OoDryefscocbqonE']",
  "event_id=$7oGQwQiBEuL1XbpVzLh4ofeIWC8OoDryefscocbqonE,depth=10,body=Historical 51 (batch=0),prevs=['$KHHHJA7Aana-fqaczESSP1CidOzlPYbe2qIyg8ewF0w']",
  "event_id=$KHHHJA7Aana-fqaczESSP1CidOzlPYbe2qIyg8ewF0w,depth=10,body=Historical 50 (batch=0),prevs=['$adP29CH4kXafp4Pm3Lq4FS7O3V4Qmu9Nka_26y6IpJE']",
  "event_id=$adP29CH4kXafp4Pm3Lq4FS7O3V4Qmu9Nka_26y6IpJE,depth=10,body=Historical 49 (batch=0),prevs=['$ZP0d_kAykGPXKYkg97MZ-tEkYf2X-LJd-Y0S71LmsD8']",
  "event_id=$ZP0d_kAykGPXKYkg97MZ-tEkYf2X-LJd-Y0S71LmsD8,depth=10,body=Historical 48 (batch=0),prevs=['$Ap4LBg6EKecDbOqsIUNarVH8M-sxMfUE4fvnHqi6oEo']",
  "event_id=$Ap4LBg6EKecDbOqsIUNarVH8M-sxMfUE4fvnHqi6oEo,depth=10,body=Historical 47 (batch=0),prevs=['$x2pHAmfPro2CaMGkbNT-ZEHbD1PQaEj6PfHHC5iDRIQ']",
  "event_id=$x2pHAmfPro2CaMGkbNT-ZEHbD1PQaEj6PfHHC5iDRIQ,depth=10,body=Historical 46 (batch=0),prevs=['$FgWKeJRbGkxaSTZGRTJZ0lV4B4irja-_gfJyUDz8M3c']",
  "event_id=$FgWKeJRbGkxaSTZGRTJZ0lV4B4irja-_gfJyUDz8M3c,depth=10,body=Historical 45 (batch=0),prevs=['$fMkY9bmk46l5PyptFHRycBdotFYxLYyA4hgEZF9pjAY']",
  "event_id=$fMkY9bmk46l5PyptFHRycBdotFYxLYyA4hgEZF9pjAY,depth=10,body=Historical 44 (batch=0),prevs=['$g4pVnEe1p2EpbOSougYf8OZgw2CeI9iByEr3uDrSPuY']",
  "event_id=$g4pVnEe1p2EpbOSougYf8OZgw2CeI9iByEr3uDrSPuY,depth=10,body=Historical 43 (batch=0),prevs=['$ojVbpac4lkvgmJo9Mx0ne6FUlo7ACg2LfoGftpUH9ZI']",
  "event_id=$ojVbpac4lkvgmJo9Mx0ne6FUlo7ACg2LfoGftpUH9ZI,depth=10,body=Historical 42 (batch=0),prevs=['$F9Hc7a0A9qxAi_DezzzKUsk13yChWE0d42fSaYBjgPc']",
  "event_id=$F9Hc7a0A9qxAi_DezzzKUsk13yChWE0d42fSaYBjgPc,depth=10,body=Historical 41 (batch=0),prevs=['$pUIxTCg76o77uMfxJjIK2weqM9T6IQ6YekivZDgnD20']",
  "event_id=$pUIxTCg76o77uMfxJjIK2weqM9T6IQ6YekivZDgnD20,depth=10,body=Historical 40 (batch=0),prevs=['$9M9JqLHmJb2VCBBKjkc9Ej5IAwG8v2n4YChD1SCKjFQ']",
  "event_id=$9M9JqLHmJb2VCBBKjkc9Ej5IAwG8v2n4YChD1SCKjFQ,depth=10,body=Historical 39 (batch=0),prevs=['$ib1BeelWRyBqFi9VdEhC153NTdqQpgGLAr07ZJj9-MQ']",
  "event_id=$ib1BeelWRyBqFi9VdEhC153NTdqQpgGLAr07ZJj9-MQ,depth=10,body=Historical 38 (batch=0),prevs=['$61R1kvPM8_Nt8rLTDdEoUAs3Ue7JpiBQmZW9de14iss']",
  "event_id=$61R1kvPM8_Nt8rLTDdEoUAs3Ue7JpiBQmZW9de14iss,depth=10,body=Historical 37 (batch=0),prevs=['$5dXix6QG4o3cZRrXitSAWE2TJ44c_9gz_pF_gQKmIZw']",
  "event_id=$5dXix6QG4o3cZRrXitSAWE2TJ44c_9gz_pF_gQKmIZw,depth=10,body=Historical 36 (batch=0),prevs=['$lEkNauoLWeP2Kc_Z0e_6FjXwDgkbzqSDxhzqYYYRsdg']",
  "event_id=$lEkNauoLWeP2Kc_Z0e_6FjXwDgkbzqSDxhzqYYYRsdg,depth=10,body=Historical 35 (batch=0),prevs=['$2BfdvXN-u6rj4JFgPP1bcr3JE9-ur7JcF7ihAYzYHUg']",
  "event_id=$2BfdvXN-u6rj4JFgPP1bcr3JE9-ur7JcF7ihAYzYHUg,depth=10,body=Historical 34 (batch=0),prevs=['$nG-fNtLkJHFSR38nZtijeabE3uYEAOuvq2HP0bUGyJk']",
  "event_id=$nG-fNtLkJHFSR38nZtijeabE3uYEAOuvq2HP0bUGyJk,depth=10,body=Historical 33 (batch=0),prevs=['$fAZc_twttB_9XybV9ajjtajqe-76chAdSdHjcz6Tfb8']",
  "event_id=$fAZc_twttB_9XybV9ajjtajqe-76chAdSdHjcz6Tfb8,depth=10,body=Historical 32 (batch=0),prevs=['$Rem6erJTsvQcpgfzWF5ia_LCwmXM7OLWvDxMltu3RB4']",
  "event_id=$Rem6erJTsvQcpgfzWF5ia_LCwmXM7OLWvDxMltu3RB4,depth=10,body=Historical 31 (batch=0),prevs=['$SQN4Q0owMx2gU-4J-qD88Tz3VXpCpLaz9n0O1bIa1kU']",
  "event_id=$SQN4Q0owMx2gU-4J-qD88Tz3VXpCpLaz9n0O1bIa1kU,depth=10,body=Historical 30 (batch=0),prevs=['$sEasprLY-iLBCFLWbZxv-MLSEaOBN0NGihSD6MM68X4']",
  "event_id=$sEasprLY-iLBCFLWbZxv-MLSEaOBN0NGihSD6MM68X4,depth=10,body=Historical 29 (batch=0),prevs=['$V8rf_7aw2xehL8qOZIwcpJfWOmphrvMTDMoFtrs950Y']",
  "event_id=$V8rf_7aw2xehL8qOZIwcpJfWOmphrvMTDMoFtrs950Y,depth=10,body=Historical 28 (batch=0),prevs=['$6BB4zKQJA6Lwz4b5xKPRuLs_ywV1dvQrAvmBCqL2YDw']",
  "event_id=$6BB4zKQJA6Lwz4b5xKPRuLs_ywV1dvQrAvmBCqL2YDw,depth=10,body=Historical 27 (batch=0),prevs=['$4AGk1DlNQhaTesircrVUXBZFDWNBq1Lut_dLJbn82JU']",
  "event_id=$4AGk1DlNQhaTesircrVUXBZFDWNBq1Lut_dLJbn82JU,depth=10,body=Historical 26 (batch=0),prevs=['$_cNNXIC9olgpnudDkug-ktUSKiZM71HvnDgPz0sWXfE']",
  "event_id=$_cNNXIC9olgpnudDkug-ktUSKiZM71HvnDgPz0sWXfE,depth=10,body=Historical 25 (batch=0),prevs=['$0ifr6p9nnJ3BCcOlxItaE8AcLdybZ5Urx-ivUJjaXHA']",
  "event_id=$0ifr6p9nnJ3BCcOlxItaE8AcLdybZ5Urx-ivUJjaXHA,depth=10,body=Historical 24 (batch=0),prevs=['$-8-I_FslO7Gvj-PL3pM9ogEgRjYV2aBrY9rLbDCgC2k']",
  "event_id=$-8-I_FslO7Gvj-PL3pM9ogEgRjYV2aBrY9rLbDCgC2k,depth=10,body=Historical 23 (batch=0),prevs=['$9r34krvdKDH9LKhcDjrFWp7daXJDwMupq-um9orXApk']",
  "event_id=$9r34krvdKDH9LKhcDjrFWp7daXJDwMupq-um9orXApk,depth=10,body=Historical 22 (batch=0),prevs=['$z7Fa5Kn9uoQLpC1gjaTbtoye5TdLWv-kx2zdXOLrC54']",
  "event_id=$z7Fa5Kn9uoQLpC1gjaTbtoye5TdLWv-kx2zdXOLrC54,depth=10,body=Historical 21 (batch=0),prevs=['$2uuOUt4M_qqs9mAMJWWK7LXPPB2FYuUal0v5iC06zIY']",
  "event_id=$2uuOUt4M_qqs9mAMJWWK7LXPPB2FYuUal0v5iC06zIY,depth=10,body=Historical 20 (batch=0),prevs=['$PRTtMhVGlUpUAHoT9p2GVnkY2GHNi-To809HFN6TkUc']",
  "event_id=$PRTtMhVGlUpUAHoT9p2GVnkY2GHNi-To809HFN6TkUc,depth=10,body=Historical 19 (batch=0),prevs=['$ANgsRYx4tP4i41cB3ZVOZTsfP3HzzjQyKACmBpAafmI']",
  "event_id=$ANgsRYx4tP4i41cB3ZVOZTsfP3HzzjQyKACmBpAafmI,depth=10,body=Historical 18 (batch=0),prevs=['$K07Lok7DO2mnuDz7rRXWf2XshAMCSkKGq9A0u-5Hp_o']",
  "event_id=$K07Lok7DO2mnuDz7rRXWf2XshAMCSkKGq9A0u-5Hp_o,depth=10,body=Historical 17 (batch=0),prevs=['$wQ6iaNkRY4vMPjzpsV8WNaqoKxyjIrIn3zUZGNtPGls']",
  "event_id=$wQ6iaNkRY4vMPjzpsV8WNaqoKxyjIrIn3zUZGNtPGls,depth=10,body=Historical 16 (batch=0),prevs=['$Z-6TMGWoXxL6USvSNUmv6asnUH4ra-YOUi0O_uNADCM']",
  "event_id=$Z-6TMGWoXxL6USvSNUmv6asnUH4ra-YOUi0O_uNADCM,depth=10,body=Historical 15 (batch=0),prevs=['$U8iKKA7TdOTva0uXFkDOuQZS9xi6GuMewMu6yVL7t-c']",
  "event_id=$U8iKKA7TdOTva0uXFkDOuQZS9xi6GuMewMu6yVL7t-c,depth=10,body=Historical 14 (batch=0),prevs=['$lOory_yjG64n9-CNSpPXcOBdJ2Hp6YHT69qb5aLhsFE']",
  "event_id=$lOory_yjG64n9-CNSpPXcOBdJ2Hp6YHT69qb5aLhsFE,depth=10,body=Historical 13 (batch=0),prevs=['$G3fi25UZRO9eEp-JbAFAmeelosmpKw_ezNY1M2hpxEc']",
  "event_id=$G3fi25UZRO9eEp-JbAFAmeelosmpKw_ezNY1M2hpxEc,depth=10,body=Historical 12 (batch=0),prevs=['$WNFtelKcHvICGdB3uhvz5VhZesHu28smlL7gT_hDlrg']",
  "event_id=$WNFtelKcHvICGdB3uhvz5VhZesHu28smlL7gT_hDlrg,depth=10,body=Historical 11 (batch=0),prevs=['$zgiMckYUJhO2DzDngM11MbNRzPewo6N21rZYb6LR9Ko']",
  "event_id=$zgiMckYUJhO2DzDngM11MbNRzPewo6N21rZYb6LR9Ko,depth=10,body=Historical 10 (batch=0),prevs=['$nnNTSnfupektlvP2WAl2Ofmuzm13pjwGvBky-SbCZPQ']",
  "event_id=$nnNTSnfupektlvP2WAl2Ofmuzm13pjwGvBky-SbCZPQ,depth=10,body=Historical 9 (batch=0),prevs=['$WvzAHtMLWzv1KVtImLt9I7qlDXKjjgFS__CyS5uw2Go']",
  "event_id=$WvzAHtMLWzv1KVtImLt9I7qlDXKjjgFS__CyS5uw2Go,depth=10,body=Historical 8 (batch=0),prevs=['$XOe653Bs1shkL8UyzxnLdvFSQSz39dZfrQnwGJaTHTs']",
  "event_id=$XOe653Bs1shkL8UyzxnLdvFSQSz39dZfrQnwGJaTHTs,depth=10,body=Historical 7 (batch=0),prevs=['$WL7ORFfraUf9pDiRMFOzwXHIkicS2Fg14RBM5JGzkcY']",
  "event_id=$WL7ORFfraUf9pDiRMFOzwXHIkicS2Fg14RBM5JGzkcY,depth=10,body=Historical 6 (batch=0),prevs=['$JeKL-CqsynXjLMunbu5OYYTFE3e_l0BpMHIP-99xOwY']",
  "event_id=$JeKL-CqsynXjLMunbu5OYYTFE3e_l0BpMHIP-99xOwY,depth=10,body=Historical 5 (batch=0),prevs=['$cLXTHBY3xCtzHjwShMqL5HWKbW0Zv9K-wBaoZv6h3ro']",
  "event_id=$cLXTHBY3xCtzHjwShMqL5HWKbW0Zv9K-wBaoZv6h3ro,depth=10,body=Historical 4 (batch=0),prevs=['$dc1-00h9Wt7xWq333J1yC_L3xQvhLhHftoioSbc_WTc']"
]
```
 
 
### Scenario 2: When the historical batch fits within the backfill chunk with "live" events before and after

But our previous silver bullet breaks when the historical batch is small and you get `eventIDsBefore` and `eventIDsAfter` all in the same backfill chunk.

Now we need a fake edge to shove the historical batch below `eventIDsBefore` (A). We can accomplish this by connecting the `batch0` insertion event -> `eventIDsBefore`(A).

The problem is that there is no straight forward path to add this fake edge. When we're at the `batch0` insertion event, nothing hints us to the fact that insertion event is related or comes before `A`. We would have to travel back down the historical chain until we hit another insertion event and check that its' `prev_event` is pointing at a non-historical event.

Also keep in mind that there could be multiple historical batches which fit in the same historical chunk so we have to travel down multiple batches to find which event we need to connect to. While all of this historical branch navigation is possible, this is getting very cumbersome. The code just to handle the previous partial bottom case is already pretty complex. I'm not too keen to fix this other problem scenario if we're just going to throw it away and decide not to do this.


When 1 chunk fits | When multiple chunks fit
--- | ---
[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBCIC0tLS0tLS0tLS0tLS0tLS0tPiBBXG4gICAgZW5kXG4gICAgXG4gICAgc3ViZ3JhcGggYmF0Y2gwXG4gICAgICAgIGJhdGNoMC1iYXRjaFtbXCJiYXRjaFwiXV0gLS0-IGJhdGNoMC0yKChcIjJcIikpIC0tPiBiYXRjaDAtMSgoMSkpIC0tPiBiYXRjaDAtMCgoMCkpIC0tPiBiYXRjaDAtaW5zZXJ0aW9uWy9pbnNlcnRpb25cXF1cbiAgICBlbmRcblxuXG4gICAgQiAtLSBmYWtlIGVkZ2UgLS0-IGJhdGNoMC1pbnNlcnRpb25CYXNlXG4gICAgYmF0Y2gwLWluc2VydGlvbkJhc2UgLS0gZmFrZSBlZGdlIC0tPiBiYXRjaDAtYmF0Y2hcbiAgICBiYXRjaDAtaW5zZXJ0aW9uIC0tIGZha2UgZWRnZSAtLS0tLS0tLS0tPiBBXG5cbiAgICBiYXRjaDAtaW5zZXJ0aW9uQmFzZVsvaW5zZXJ0aW9uXFxdIC0tLS0tLS0tLS0tLS0tLT4gQVxuXG4iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ) ![](https://user-images.githubusercontent.com/558581/139520655-d0b5dd0e-9a94-44cc-bbae-5138d90f1818.png) | [Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBCIC0tLS0tLS0tLS0tLS0tLS0tLT4gQVxuICAgIGVuZFxuICAgIFxuICAgIHN1YmdyYXBoIGJhdGNoMFxuICAgICAgICBiYXRjaDAtYmF0Y2hbW1wiYmF0Y2hcIl1dIC0tPiBiYXRjaDAtMigoXCIyXCIpKSAtLT4gYmF0Y2gwLTEoKDEpKSAtLT4gYmF0Y2gwLTAoKDApKSAtLT4gYmF0Y2gwLWluc2VydGlvblsvaW5zZXJ0aW9uXFxdXG4gICAgZW5kXG5cbiAgICBzdWJncmFwaCBiYXRjaDFcbiAgICAgICAgYmF0Y2gxLWJhdGNoW1tcImJhdGNoXCJdXSAtLT4gYmF0Y2gxLTIoKFwiMlwiKSkgLS0-IGJhdGNoMS0xKCgxKSkgLS0-IGJhdGNoMS0wKCgwKSkgLS0-IGJhdGNoMS1pbnNlcnRpb25bL2luc2VydGlvblxcXVxuICAgIGVuZFxuICAgIFxuICAgIHN1YmdyYXBoIGJhdGNoMlxuICAgICAgICBiYXRjaDItYmF0Y2hbW1wiYmF0Y2hcIl1dIC0tPiBiYXRjaDItMigoXCIyXCIpKSAtLT4gYmF0Y2gyLTEoKDEpKSAtLT4gYmF0Y2gyLTAoKDApKSAtLT4gYmF0Y2gyLWluc2VydGlvblsvaW5zZXJ0aW9uXFxdXG4gICAgZW5kXG5cbiAgICBCIC0tIGZha2UgZWRnZSAtLT4gYmF0Y2gwLWluc2VydGlvbkJhc2VcbiAgICBiYXRjaDAtaW5zZXJ0aW9uQmFzZSAtLSBmYWtlIGVkZ2UgLS0-IGJhdGNoMC1iYXRjaFxuICAgIGJhdGNoMi1pbnNlcnRpb24gLS0gZmFrZSBlZGdlIC0tLS0-IEFcblxuICAgIGJhdGNoMC1pbnNlcnRpb25CYXNlWy9pbnNlcnRpb25cXF0gLS0tLS0tLS0tLS0tLS0tPiBBXG4gICAgYmF0Y2gxLWJhdGNoIC0uLT4gYmF0Y2gwLWluc2VydGlvblxuICAgIGJhdGNoMi1iYXRjaCAtLi0-IGJhdGNoMS1pbnNlcnRpb25cbiIsIm1lcm1haWQiOiJ7XG4gIFwidGhlbWVcIjogXCJkZWZhdWx0XCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9) ![](https://user-images.githubusercontent.com/558581/139520963-5400da47-2b52-4e53-9dab-8831def8989c.png)

Reply:
The main reason we're going to opt to not sort topologically when receiving backfill events is because it's probably best to do whatever is easiest to make it just work. People will probably have opinions once they look at [MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716) which could change whatever implementation anyway.

As mentioned, ideally we would do this but code necessary to make the fake edges gets confusing and gives an impression of “just whyyyy” (feels icky). This problem also dissolves with online topological ordering.